<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Bastien CHASSAGNOL and Savandara BESSE">
<meta name="dcterms.date" content="2024-11-28">

<title>download_datasets_and_metadata – Reference profiles for Drosophila deconvolution</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Reference profiles for Drosophila deconvolution</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./01_download_datasets_and_metadata.html">Download datasets</a></li><li class="breadcrumb-item"><a href="./01_download_datasets_and_metadata.html">Tidyverse libraries</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Download datasets</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_download_datasets_and_metadata.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Tidyverse libraries</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Generate reference profiles</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_02_preprocess_scRNASeq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Relevance proof</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#protocol-to-downnload-and-retrieve-metadata-to-build-reference-datasets" id="toc-protocol-to-downnload-and-retrieve-metadata-to-build-reference-datasets" class="nav-link active" data-scroll-target="#protocol-to-downnload-and-retrieve-metadata-to-build-reference-datasets">Protocol to downnload and retrieve metadata to build reference datasets</a>
  <ul class="collapse">
  <li><a href="#litterature-and-web-scraping" id="toc-litterature-and-web-scraping" class="nav-link" data-scroll-target="#litterature-and-web-scraping">Litterature and web-scraping</a>
  <ul class="collapse">
  <li><a href="#optionnal-retrieve-programmatically-drscdb-single-cell-databases" id="toc-optionnal-retrieve-programmatically-drscdb-single-cell-databases" class="nav-link" data-scroll-target="#optionnal-retrieve-programmatically-drscdb-single-cell-databases">(Optionnal) Retrieve programmatically DRscDB single cell databases</a></li>
  </ul></li>
  <li><a href="#retrieve-phenotype-annotation-for-each-dataset" id="toc-retrieve-phenotype-annotation-for-each-dataset" class="nav-link" data-scroll-target="#retrieve-phenotype-annotation-for-each-dataset">Retrieve phenotype annotation for each dataset</a>
  <ul class="collapse">
  <li><a href="#modencode-metadata" id="toc-modencode-metadata" class="nav-link" data-scroll-target="#modencode-metadata">The ModENCODE pheno data</a></li>
  <li><a href="#the-flyatlas2-pheno-data" id="toc-the-flyatlas2-pheno-data" class="nav-link" data-scroll-target="#the-flyatlas2-pheno-data">The FlyAtlas2 pheno data</a></li>
  <li><a href="#sec-convert-sql-format" id="toc-sec-convert-sql-format" class="nav-link" data-scroll-target="#sec-convert-sql-format">(Optionnal) Convert old SQL version 5 to modern SQL version 9</a></li>
  <li><a href="#sec-fca-pheno-data" id="toc-sec-fca-pheno-data" class="nav-link" data-scroll-target="#sec-fca-pheno-data">The single cell Fly Cell Atlas pheno data</a></li>
  </ul></li>
  <li><a href="#retrieve-transcriptomic-annotation" id="toc-retrieve-transcriptomic-annotation" class="nav-link" data-scroll-target="#retrieve-transcriptomic-annotation">Retrieve transcriptomic annotation</a>
  <ul class="collapse">
  <li><a href="#sec-modencode-expression" id="toc-sec-modencode-expression" class="nav-link" data-scroll-target="#sec-modencode-expression">The ModENCODE expression matrix</a></li>
  <li><a href="#the-flyatlas2-expression-matrix" id="toc-the-flyatlas2-expression-matrix" class="nav-link" data-scroll-target="#the-flyatlas2-expression-matrix">The FlyAtlas2 expression matrix</a></li>
  <li><a href="#the-fly-cell-atlas-expression-matrix" id="toc-the-fly-cell-atlas-expression-matrix" class="nav-link" data-scroll-target="#the-fly-cell-atlas-expression-matrix">The Fly Cell Atlas expression matrix</a></li>
  </ul></li>
  </ul></li>
  
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./01_download_datasets_and_metadata.html">Download datasets</a></li><li class="breadcrumb-item"><a href="./01_download_datasets_and_metadata.html">Tidyverse libraries</a></li></ol></nav>
<div class="quarto-title">
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Bastien CHASSAGNOL and Savandara BESSE </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 28, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tidyverse libraries</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(knitr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr) <span class="co"># For regex operations</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr) <span class="co"># vectorial operations</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr) <span class="co"># read and write tsv and csv files</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co">#  For website exploration</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rvest)  <span class="co"># For web scraping and HTML parsing</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(jsonlite)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(httr2) <span class="co"># tailored for HTTP/HTTPS requests</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co"># both packages are better tailored to FTP protocols</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curl) </span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RCurl)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co"># For SQL parsing and integration with R</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DBI) <span class="co"># open connection with DBMS</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dbplyr) <span class="co"># seamless use of dplyr operations with SQL power</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RSQLite) <span class="co"># lightweight DBMS</span></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Set global options</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>opts_chunk<span class="sc">$</span><span class="fu">set</span>(</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">echo =</span> <span class="cn">TRUE</span>,         <span class="co"># Show R code in the output</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">eval =</span> <span class="cn">FALSE</span>,        <span class="co"># By default, do not execute R code</span></span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">warning =</span> <span class="cn">FALSE</span>,     <span class="co"># Suppress warnings in the output</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">message =</span> <span class="cn">FALSE</span>,     <span class="co"># Suppress messages in the output</span></span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">fig.width =</span> <span class="dv">7</span>,       <span class="co"># Default figure width</span></span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">fig.height =</span> <span class="dv">5</span>,      <span class="co"># Default figure height</span></span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">fig.align =</span> <span class="st">"center"</span> <span class="co"># Centre-align figures</span></span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Set ggplot2 minimal theme for consistency</span></span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="protocol-to-downnload-and-retrieve-metadata-to-build-reference-datasets" class="level1">
<h1>Protocol to downnload and retrieve metadata to build reference datasets</h1>
<section id="litterature-and-web-scraping" class="level2">
<h2 class="anchored" data-anchor-id="litterature-and-web-scraping">Litterature and web-scraping</h2>
<ul>
<li>RNASeq, tissue-resolved datasets:
<ul>
<li>the <a href="https://wiki.flybase.org/wiki/FlyBase:Drosophila_Online_Resources#Gene_Expression_Databases_and_Tools">FlyBase:Drosophila_Online_Resources</a></li>
<li>The Wiki fly page also refers to a a list of tools, Web APIs and curated pipelines to analyse Drosophila samples.</li>
<li>Most popular and relevant studies are <a href="https://flybase.org/rnaseq/profile_search"><code>modENCODE</code> tissue and cell lines experiences</a> and <a href="https://academic.oup.com/nar/article/46/D1/D809/4563305"><code>FlyAtlas2</code></a><a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.<br>
</li>
</ul></li>
<li>scRNASeq datasets:
<ul>
<li><a href="https://wiki.flybase.org/wiki/FlyBase:Drosophila_Online_Resources#Single_Cell_RNA-seq">List of scRNASeq datasets and tools, at the tissue level</a></li>
<li><a href="https://www.flyrnai.org/tools/single_cell/web/summary">Comprehensive list of scRNASeq databases, within a tissue level, using the <code>DRscDB</code> Shiny App</a></li>
</ul></li>
<li><a href="https://flybase.org/downloads/bulkdata">Complete list of Datasets, from FlyAtlas compendium</a></li>
</ul>
<p><strong>All these datasets, with some metadata characterising them, are reported in Excel table <code>data/databases_metadata.xlsx</code></strong>, composed of three sheets:</p>
<ol type="1">
<li><code>Databases references</code>, enumerating the most relevant datasets and RNASeq resources.</li>
<li><code>DRscDB Databases Single Cell</code>, listing all scRNASeq datasets having a higher granularity than tissue level (for instance, characterising the cell subtypes composing the digestive tract).</li>
</ol>
<section id="optionnal-retrieve-programmatically-drscdb-single-cell-databases" class="level3">
<h3 class="anchored" data-anchor-id="optionnal-retrieve-programmatically-drscdb-single-cell-databases">(Optionnal) Retrieve programmatically DRscDB single cell databases</h3>
<ol type="1">
<li>Toggle the <a href="https://www.flyrnai.org/tools/single_cell/web/summary"><code>DRscDB</code> Datasets tab</a>.</li>
<li>Retrieve the metadata dataset using the R web-scraping tool <code>rvest</code>, and select only the most relevant columns:</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>url_DRscDB <span class="ot">&lt;-</span> <span class="st">"https://www.flyrnai.org/tools/single_cell/web/summary"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>irrelevant_columns_DRscDB <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"pubmed date"</span>, <span class="st">"dataset ids"</span>, <span class="st">"subset names"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"species"</span>, <span class="st">"species id"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"reads cell"</span>, <span class="st">"genes cell"</span>, <span class="st">"total gene"</span>,</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"batch correction"</span>, <span class="st">"data in supp table"</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"web link available"</span>, <span class="st">"linkout URL"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>metadata_DRscDB <span class="ot">&lt;-</span> <span class="fu">read_html</span>(url_DRscDB) <span class="sc">|&gt;</span> <span class="co"># read a HTML file</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">html_table</span>() <span class="sc">|&gt;</span> <span class="co"># from the HTML file, select only Tables (on that website, only one found)</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">pluck</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(species <span class="sc">%in%</span> <span class="st">"Drosophila"</span>) <span class="sc">|&gt;</span> <span class="co"># Filter on Drosophila species</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">all_of</span>(irrelevant_columns_DRscDB))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="retrieve-phenotype-annotation-for-each-dataset" class="level2">
<h2 class="anchored" data-anchor-id="retrieve-phenotype-annotation-for-each-dataset">Retrieve phenotype annotation for each dataset</h2>
<div class="callout callout-style-default callout-warning callout-titled" title="Major issues for retrieving Phenotypes">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Major issues for retrieving Phenotypes
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>No homogeneous database storage (SQL, Excel, SRA, tsv file, …), within and in-between studies</li>
<li>While the FlyBase database is comprehensive, does not store properly annotation data for each tissue, at least for RNASeq samples<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</li>
<li>No official and up-to-date BioConductor nor CRAN package to fetch and explore Drosophila databases</li>
</ul>
<p><strong>Consequence</strong>: specific protocol for retrieving such phenotypical annotation for each dataset (I enumerate for the top three popular ones).</p>
</div>
</div>
<section id="modencode-metadata" class="level3">
<h3 class="anchored" data-anchor-id="modencode-metadata">The ModENCODE pheno data</h3>
<ul>
<li>The easiest way I could find: retrieve metadata from <a href="https://wiki.flybase.org/wiki/FlyBase:ModENCODE_data_at_FlyBase#RNA-Seq_JBrowse_Track_Listing">JBrowse FlyBase Wiki website</a></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ModENCODE_databases <span class="ot">&lt;-</span> <span class="fu">read_html</span>(<span class="st">"https://wiki.flybase.org/wiki/FlyBase:ModENCODE_data_at_FlyBase#RNA-Seq_JBrowse_Track_Listing"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">html_table</span>() <span class="sc">|&gt;</span> <span class="co"># `trim = FALSE` to ensure missing cells are filled correctly </span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">pluck</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span> </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="st">`</span><span class="at">Track section</span><span class="st">`</span> <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Expression&gt;RNA-Seq &gt;modENCODE Transcriptomes  &gt;Tissues"</span>))</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>ModENCODE_databases_formatted <span class="ot">&lt;-</span> ModENCODE_databases <span class="sc">|&gt;</span> </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(stringr<span class="sc">::</span><span class="fu">str_detect</span>(<span class="st">`</span><span class="at">Track section</span><span class="st">`</span>, <span class="st">"Tissues$"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="st">`</span><span class="at">Track section</span><span class="st">`</span>, <span class="sc">-</span> <span class="st">"JBrowse Track Description"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">tissue =</span> <span class="st">"Track name"</span>, <span class="at">FlybaseID=</span><span class="st">"FlyBase Dataset(s)"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">FlybaseID =</span> <span class="fu">lapply</span>(stringr<span class="sc">::</span><span class="fu">str_split</span>(FlybaseID,        <span class="co"># split by mE_mRNA, without consuming the separator</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">pattern =</span> <span class="st">"(?=mE_mRNA)"</span>), str_subset, <span class="st">"[^ ]"</span>)) <span class="sc">|&gt;</span>  <span class="co"># remove empty chains </span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">unchop</span>(FlybaseID) <span class="co"># from compact list of samples per tissue type to tidy format</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">write_csv</span>(ModENCODE_databases_formatted, </span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"./data/RNASeq/ModENCODE_pheno_data.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Alternative methods:
<ul>
<li>When pasting a HTML table to an Excel Workbook, use <strong>Paste with a refreshable web query</strong> option</li>
<li>The <a href="https://flybase.org/rnaseq/profile_search">FlyBase RNASeq Search profile</a>, clicking on the <code>tissue</code> and optionally <code>cell line</code> items.</li>
</ul></li>
</ul>
<div>

</div>
<div class="quarto-layout-panel" data-layout="[[1], [1,1]]">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 100.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/FlyBase_Profile_Search.png" class="img-fluid figure-img"></p>
<figcaption>Screenshot FlyBase Profile Search</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/FlyBase_tissues.png" class="img-fluid figure-img"></p>
<figcaption>Screenshot Tissue Samples</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/FlyBase_cell_lines.png" class="img-fluid figure-img"></p>
<figcaption>Screenshot Cell Line samples</figcaption>
</figure>
</div>
</div>
</div>
</div>
<ul>
<li>To finalise the annotation, we’re going to use the <a href="https://flybase.github.io/api/swagger-ui/#/HitList/getFlyBaseHitList">Rest API GET HitList</a>, programmatic access enabled by the FlyBase website:</li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Which individual experiences we want metadata from?</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>flybase_ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"FBlc0000215"</span>, <span class="st">"FBlc0000216"</span>, <span class="st">"FBlc0000217"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a URL with IDs delimited by commas</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>base_url <span class="ot">&lt;-</span> <span class="st">"https://api.flybase.org/api/v1.0/hitlist/fetch"</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>modENCODE_request <span class="ot">&lt;-</span> <span class="fu">request</span>(base_url) <span class="sc">|&gt;</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># parameter construction defined here: https://flybase.github.io/api/swagger-ui/#/HitList/getFlyBaseHitList</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(flybase_ids <span class="sc">|&gt;</span> </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">paste</span>(<span class="at">collapse =</span> <span class="st">","</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>()</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the response status, 200 corresponding to a success</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">resp_status</span>(modENCODE_request) <span class="sc">==</span> <span class="dv">200</span> <span class="sc">&amp;</span> <span class="fu">resp_has_body</span>(modENCODE_request)) {</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Parse the JSON content into an R object</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  ModENCODE_metadata <span class="ot">&lt;-</span>   modENCODE_request <span class="sc">|&gt;</span> </span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">resp_body_string</span>() <span class="sc">|&gt;</span> </span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">fromJSON</span>()  <span class="sc">|&gt;</span> </span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">pluck</span>(<span class="st">"resultset"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    purrr<span class="sc">::</span><span class="fu">pluck</span>(<span class="st">"result"</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># convert a HTTP error to a R error</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resp_check_status</span>(modENCODE_request)</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="the-flyatlas2-pheno-data" class="level3">
<h3 class="anchored" data-anchor-id="the-flyatlas2-pheno-data">The FlyAtlas2 pheno data</h3>
<div class="callout callout-style-default callout-warning callout-titled" title="Excel workbook and aggregated dataset">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Excel workbook and aggregated dataset
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <a href="https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help">Excel Workbook</a> mentioned in the <code>Docs</code> tab section of the official <code>FlyAtlas2</code> initiative only stores aggregated and averaged information at the tissue level, rendering it useless for proper downstream analyses<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>.</p>
<p><strong>In conclusion, this method and this dataset is irrelevant for most downstream analyses</strong><a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a></p>
</div>
</div>
<ul>
<li>In the FPKM/RPKM file stored in the web encyclopaedia <code>FlyBase</code>, the column headers identifying uniquely each sample describe the developmental stage, the sex and the original tissue. However, it’s not really practical to parse, nor follows an unique identifier composition. Example: <code>RNA-Seq_Profile_FlyAtlas2_Adult_Female_Brain_(FBlc0003619)</code>.
<ul>
<li>However, we can use the same REST API trick mentioned in <span class="citation" data-cites="API-FlyBase-modENCODE-metadata">(<a href="#ref-API-FlyBase-modENCODE-metadata" role="doc-biblioref"><strong>API-FlyBase-modENCODE-metadata?</strong></a>)</span> to get some additional information (column <code>title</code>).</li>
</ul></li>
</ul>
<ul>
<li><p>The most challenging method involves re-running the SQL file provided with the project.</p>
<ol type="1">
<li>The name of the SQL file can be retrieved on the <a href="https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help"><code>FlyAtlas2</code> website</a>, or, in a somehow twisted way with the following R code<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>.</li>
</ol></li>
</ul>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>flyatlas_url <span class="ot">&lt;-</span> <span class="st">"https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help"</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>webpage_text <span class="ot">&lt;-</span> <span class="fu">read_html</span>(flyatlas_url) <span class="sc">|&gt;</span> </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">html_text</span>()</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>flyatlas_files <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_extract_all</span>(webpage_text, </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                                         <span class="st">"motif.mvls.gla.ac.uk/downloads/[^</span><span class="sc">\\</span><span class="st">s]+"</span>, <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> <span class="fu">as.vector</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>From the previous manipulation, two files are available: motif.mvls.gla.ac.uk/downloads/FlyAtlas2_2024.01.08.sql, motif.mvls.gla.ac.uk/downloads/FlyAtlas2_gene_data.xlsx, but only the SQL file is comprehensive, storing gene expressions at the individual level (and not at the aggregated tissue level). i. The SQL relational database is described in details in <a href="https://pmc.ncbi.nlm.nih.gov/articles/instance/5753349/bin/gkx976_supp.pdf">Supplementary FlyAtlas 2 documents, Table S2, Pages 3-4</a>. Of interest, the <code>Tissue</code>, <code>GeneFPKM</code> and <code>GeneRPM</code> datasets<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>. ii. Parse the SQL file’s instructions, keeping SQL lines for building <code>Tissue</code>, <code>GeneFPKM</code> and <code>GeneRPM</code> datasets. Usually, building a SQL dataset involves three stages: ensuring the non prior existence of the dataset (if it’s the case, enforce its deletion), define the dataset (name and type of features/columns), and finally, build the dataset itself (indeed, dataset content is hard-coded within the SQL file):</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sql_instructions <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_lines</span>(<span class="st">'data/FlyAtlas2_2024.01.08.sql'</span>,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">skip_empty_rows =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Keep only lines that are NOT empty nor commented (defined by "--")</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># stringr::str_subset("^\\s*(--)|/\\*", negate = TRUE) |&gt;</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_subset</span>(<span class="st">"^</span><span class="sc">\\</span><span class="st">s*(--)|/</span><span class="sc">\\</span><span class="st">*|(UN)?LOCK"</span>, <span class="at">negate =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(<span class="at">collapse =</span> <span class="st">"</span><span class="sc">\r\n</span><span class="st">"</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_extract_all</span>(<span class="at">pattern =</span> <span class="fu">regex</span>(<span class="st">"[^;]+;"</span>, <span class="at">multiline =</span> <span class="cn">TRUE</span>,<span class="at">dotall =</span> <span class="cn">TRUE</span>),</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                  <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.vector</span>()</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>tissue_index <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"CREATE TABLE `Tissue`"</span>, sql_instructions, <span class="at">fixed =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>tissue_instructions <span class="ot">&lt;-</span> sql_instructions[(tissue_index<span class="dv">-1</span>)<span class="sc">:</span>(tissue_index <span class="sc">+</span> <span class="dv">1</span>)]</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">write_lines</span>(tissue_instructions,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"data/old_tissue_parsing.sql"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="i">
<li>Use modern SQL DBMS. Report to section <a href="#sec-convert-sql-format" class="quarto-xref">Section&nbsp;1.2.3</a> for further details. Major issue with the current Fly Atlas 2 database is their outdated use of the <code>myISAM</code> DBMS standards, for which no official converter exists.<br>
</li>
<li>(Optional) Verify your updated SQL instructions can be parsed and executed using your local DBMS manager like mySQL, or even simpler, going on <a href="https://www.programiz.com/sql/online-compiler/">SQL online resources</a>. It turned out that we had to remove an uniqueness constraint on <code>Tissue.Abbreviation</code> feature to run the SQL code.</li>
<li>Execute the modern SQL file, store transiently the output to a SQLite database, collect and export the results as a standard csv file.</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create SQLite database connection</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>con <span class="ot">&lt;-</span> <span class="fu">dbConnect</span>(RSQLite<span class="sc">::</span><span class="fu">SQLite</span>(), <span class="st">"data/FlyAtlas2_tissue.sqlite"</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>sql_statements <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_lines</span>(<span class="st">"data/new_tissue.sql"</span>, </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">skip_empty_rows =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_subset</span>(<span class="st">"^</span><span class="sc">\\</span><span class="st">s*(--)|/</span><span class="sc">\\</span><span class="st">*|(UN)?LOCK"</span>, <span class="at">negate =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(<span class="at">collapse =</span> <span class="st">"</span><span class="sc">\r\n</span><span class="st">"</span>) <span class="sc">|&gt;</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_extract_all</span>(<span class="at">pattern =</span> <span class="fu">regex</span>(<span class="st">"[^;]+;"</span>, <span class="at">multiline =</span> <span class="cn">TRUE</span>,<span class="at">dotall =</span> <span class="cn">TRUE</span>),</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.vector</span>()</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Tip: since dbExecute can only perform one query after the other, we use a for-loop to perform sequential SQL queries</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (statement <span class="cf">in</span> sql_statements) {</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dbExecute</span>(con, statement)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a><span class="co"># testthat::expect_contains(dbListTables(con), "Tissue")</span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="co"># DBI::dbRemoveTable(con, "Tissue")</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Export the tissue dataset to a standard CSV format</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>tissue_fly_atlas <span class="ot">&lt;-</span> <span class="fu">tbl</span>(con, <span class="st">"Tissue"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">collect</span>()</span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">write_csv</span>(tissue_fly_atlas,</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"data/pheno_data_flyatlas2.csv"</span>)</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Close the connection once done</span></span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true" tabindex="-1"></a><span class="fu">dbDisconnect</span>(con)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sec-convert-sql-format" class="level3">
<h3 class="anchored" data-anchor-id="sec-convert-sql-format">(Optionnal) Convert old SQL version 5 to modern SQL version 9</h3>
<p>While functionalities going along the <code>DBI</code> R package would theoretically enable us to execute SQL instructions directly from R, it turns out that the SQL instructions are tailored to <a href="https://en.wikipedia.org/wiki/MyISAM">MyISAM DBMS software</a>, outdated from 2009, and can not be processed directly by the <code>DBI</code> wrapper. Besides, <code>DBI</code> works best with <code>SQLite</code> DBMS, and the FlyBase SQL file has not been written to that end. To conclude, we have to convert the SQL FlyBase file to modern SQL standards (current version: <code>SQL Server 2022</code>) while ensuring compatibility with <code>SQLite</code>.</p>
<p>No dedicated tools, up to my knowledge, have been developed in R to that end. So definitely, the best method, while being intrinsically stochastic, would consist of pairing a LLM using API REST integration with <em>prompt engineering</em>. I propose two code snippets below with a personalised engineered prompt, using <a href="https://github.com/mlverse/chattr"><code>chattr</code></a>, see <a href="#lst-chattr-sql" class="quarto-xref">Listing&nbsp;1</a> and <a href="https://www.r-bloggers.com/2024/06/large-language-models-llms-in-r-with-tidychatmodels/"><code>tidychatmodels</code></a>, see <a href="#lst-tidychatmodels-sql" class="quarto-xref">Listing&nbsp;2</a>, respectively<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>.</p>
<div class="cell" data-layout-align="center">
<div id="lst-chattr-sql" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-chattr-sql-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;1: Use chattr for updating SQL file.
</figcaption>
<div aria-describedby="lst-chattr-sql-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-chattr-sql"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-chattr-sql-1"><a href="#lst-chattr-sql-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remotes::install_github("mlverse/chattr")</span></span>
<span id="lst-chattr-sql-2"><a href="#lst-chattr-sql-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(chattr)</span>
<span id="lst-chattr-sql-3"><a href="#lst-chattr-sql-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-chattr-sql-4"><a href="#lst-chattr-sql-4" aria-hidden="true" tabindex="-1"></a><span class="co"># parameter configuration and saving</span></span>
<span id="lst-chattr-sql-5"><a href="#lst-chattr-sql-5" aria-hidden="true" tabindex="-1"></a><span class="fu">chattr_use</span>(<span class="st">"copilot"</span>)</span>
<span id="lst-chattr-sql-6"><a href="#lst-chattr-sql-6" aria-hidden="true" tabindex="-1"></a><span class="fu">chattr_defaults</span>(<span class="at">max_data_files =</span> <span class="dv">10</span>, </span>
<span id="lst-chattr-sql-7"><a href="#lst-chattr-sql-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">include_doc_contents =</span> <span class="cn">TRUE</span>, </span>
<span id="lst-chattr-sql-8"><a href="#lst-chattr-sql-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">max_data_frames =</span> <span class="dv">10</span>,</span>
<span id="lst-chattr-sql-9"><a href="#lst-chattr-sql-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">yaml_file =</span> <span class="st">"vignettes/chattr.yml"</span>)</span>
<span id="lst-chattr-sql-10"><a href="#lst-chattr-sql-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-chattr-sql-11"><a href="#lst-chattr-sql-11" aria-hidden="true" tabindex="-1"></a>new_sql <span class="ot">&lt;-</span> <span class="fu">chattr</span>(<span class="at">prompt =</span> <span class="st">"Please convert the SQL file &lt;old-sql-location&gt;, written with SQL version 5, to SQL Server 2022 syntax, while making it compliant with the SQLite DBMS constraints."</span>,</span>
<span id="lst-chattr-sql-12"><a href="#lst-chattr-sql-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">prompt_build =</span> <span class="st">"Specifically, focus on the following: Change data types to match SQLite standards (e.g., TINYINT to INTEGER, ENUM to TEXT). Replace AUTO_INCREMENT with AUTOINCREMENT. Remove any unsupported SQL syntax in SQLite. Write the resulting modernised SQL file to `data/new_tissue_global.sql`"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<div class="cell" data-layout-align="center">
<div id="lst-tidychatmodels-sql" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-tidychatmodels-sql-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;2: Use tidychatmodels for updating SQL files.
</figcaption>
<div aria-describedby="lst-tidychatmodels-sql-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-tidychatmodels-sql"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-tidychatmodels-sql-1"><a href="#lst-tidychatmodels-sql-1" aria-hidden="true" tabindex="-1"></a><span class="co"># devtools::install_github("AlbertRapp/tidychatmodels")</span></span>
<span id="lst-tidychatmodels-sql-2"><a href="#lst-tidychatmodels-sql-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidychatmodels)</span>
<span id="lst-tidychatmodels-sql-3"><a href="#lst-tidychatmodels-sql-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-tidychatmodels-sql-4"><a href="#lst-tidychatmodels-sql-4" aria-hidden="true" tabindex="-1"></a>chat_openai <span class="ot">&lt;-</span> <span class="fu">create_chat</span>(<span class="st">'openai'</span>, <span class="fu">Sys.getenv</span>(<span class="st">'OAI_DEV_KEY'</span>))<span class="sc">|&gt;</span></span>
<span id="lst-tidychatmodels-sql-5"><a href="#lst-tidychatmodels-sql-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_model</span>(<span class="st">'gpt-3.5-turbo'</span>) <span class="sc">|&gt;</span></span>
<span id="lst-tidychatmodels-sql-6"><a href="#lst-tidychatmodels-sql-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add_params(temperature = 0.5, max_tokens = 100) |&gt;</span></span>
<span id="lst-tidychatmodels-sql-7"><a href="#lst-tidychatmodels-sql-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># define the general AI agent system</span></span>
<span id="lst-tidychatmodels-sql-8"><a href="#lst-tidychatmodels-sql-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_message</span>(</span>
<span id="lst-tidychatmodels-sql-9"><a href="#lst-tidychatmodels-sql-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">role =</span> <span class="st">'system'</span>,</span>
<span id="lst-tidychatmodels-sql-10"><a href="#lst-tidychatmodels-sql-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">message =</span> <span class="st">'You are a chatbot that only returns SQL code, whose syntax is</span></span>
<span id="lst-tidychatmodels-sql-11"><a href="#lst-tidychatmodels-sql-11" aria-hidden="true" tabindex="-1"></a><span class="st">    compatible with SQL Server 2022 and SQLite DBMS.'</span></span>
<span id="lst-tidychatmodels-sql-12"><a href="#lst-tidychatmodels-sql-12" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">|&gt;</span></span>
<span id="lst-tidychatmodels-sql-13"><a href="#lst-tidychatmodels-sql-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create your customised prompt</span></span>
<span id="lst-tidychatmodels-sql-14"><a href="#lst-tidychatmodels-sql-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_message</span>(</span>
<span id="lst-tidychatmodels-sql-15"><a href="#lst-tidychatmodels-sql-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">role =</span> <span class="st">'user'</span>,</span>
<span id="lst-tidychatmodels-sql-16"><a href="#lst-tidychatmodels-sql-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">message =</span> <span class="st">'Convert old sql file &lt;old-sql-path&gt; to recent SQL code,</span></span>
<span id="lst-tidychatmodels-sql-17"><a href="#lst-tidychatmodels-sql-17" aria-hidden="true" tabindex="-1"></a><span class="st">    write the output to &lt;new-sql-path&gt;. Specifically: Update data types to align</span></span>
<span id="lst-tidychatmodels-sql-18"><a href="#lst-tidychatmodels-sql-18" aria-hidden="true" tabindex="-1"></a><span class="st">    with SQLite standards, such as replacing TINYINT with INTEGER,</span></span>
<span id="lst-tidychatmodels-sql-19"><a href="#lst-tidychatmodels-sql-19" aria-hidden="true" tabindex="-1"></a><span class="st">    ENUM with TEXT, and other incompatible types. Change AUTO_INCREMENT to</span></span>
<span id="lst-tidychatmodels-sql-20"><a href="#lst-tidychatmodels-sql-20" aria-hidden="true" tabindex="-1"></a><span class="st">    AUTOINCREMENT for primary keys. Remove syntax elements unsupported in SQLite,</span></span>
<span id="lst-tidychatmodels-sql-21"><a href="#lst-tidychatmodels-sql-21" aria-hidden="true" tabindex="-1"></a><span class="st">    like full-text indexes, foreign key constraints outside PRAGMA, or stored procedures.</span></span>
<span id="lst-tidychatmodels-sql-22"><a href="#lst-tidychatmodels-sql-22" aria-hidden="true" tabindex="-1"></a><span class="st">    Where conflicts between SQL Server and SQLite arise, prioritize SQLite compliance.</span></span>
<span id="lst-tidychatmodels-sql-23"><a href="#lst-tidychatmodels-sql-23" aria-hidden="true" tabindex="-1"></a><span class="st">    Output the transformed SQL file as a set of discrete statements, one per line,</span></span>
<span id="lst-tidychatmodels-sql-24"><a href="#lst-tidychatmodels-sql-24" aria-hidden="true" tabindex="-1"></a><span class="st">    removing empty or commented lines for clarity. Do not add any empty lines nor comments. Do not add uniqueness constraints. Convert the entire SQL file, without using further rows argument.'</span></span>
<span id="lst-tidychatmodels-sql-25"><a href="#lst-tidychatmodels-sql-25" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="lst-tidychatmodels-sql-26"><a href="#lst-tidychatmodels-sql-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-tidychatmodels-sql-27"><a href="#lst-tidychatmodels-sql-27" aria-hidden="true" tabindex="-1"></a>chat_user_result <span class="ot">&lt;-</span> chat_openai <span class="sc">|&gt;</span></span>
<span id="lst-tidychatmodels-sql-28"><a href="#lst-tidychatmodels-sql-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">perform_chat</span>()<span class="sc">|&gt;</span></span>
<span id="lst-tidychatmodels-sql-29"><a href="#lst-tidychatmodels-sql-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">extract_chat</span>(<span class="at">silent =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<p>And finally, an alternative SQL method, which mixes a bunch of regex expressions and <code>tidyverse</code> manipulations, while guaranteeing a determined output:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>tissue_colnames <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"CREATE TABLE `Tissue`"</span>, sql_instructions,</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">fixed =</span> <span class="cn">TRUE</span>, <span class="at">value =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_extract_all</span>(<span class="at">pattern =</span> <span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">`)[[:alnum:]]+(?=</span><span class="sc">\\</span><span class="st">`)"</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.vector</span>() <span class="sc">|&gt;</span> </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>() <span class="sc">|&gt;</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setdiff</span>(<span class="st">"Tissue"</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>tissue_inputs <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"INSERT INTO `Tissue`"</span>, sql_instructions,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">fixed =</span> <span class="cn">TRUE</span>, <span class="at">value =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_extract_all</span>(<span class="st">"(?&lt;=</span><span class="sc">\\</span><span class="st">()([^)]+)(?=</span><span class="sc">\\</span><span class="st">),*)"</span>,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.vector</span>() <span class="sc">|&gt;</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_replace_all</span>(<span class="st">"'"</span>, <span class="st">""</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">map</span>(\(x) stringr<span class="sc">::</span><span class="fu">str_split</span>(x, <span class="st">","</span>, <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>               tibble<span class="sc">::</span><span class="fu">as_tibble</span>()) <span class="sc">|&gt;</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">list_rbind</span>() <span class="sc">|&gt;</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setNames</span>(<span class="at">nm =</span> tissue_colnames)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">write_csv</span>(tissue_inputs,</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>                 <span class="at">file =</span> <span class="st">"data/pheno_data_flyatlas2.csv"</span>,</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>                 <span class="at">escape =</span> <span class="st">"none"</span>, <span class="at">quote =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sec-fca-pheno-data" class="level3">
<h3 class="anchored" data-anchor-id="sec-fca-pheno-data">The single cell Fly Cell Atlas pheno data</h3>
<p>One possibility to retrieve phenotype data would consist of using the <a href="https://www.bioconductor.org/packages/release/bioc/manuals/ArrayExpress/man/ArrayExpress.pdf"><code>BioConductor ArrayExpress</code></a> package. Unfortunately, <strong>it does not seem updated anymore, and was indeed unable to download and parse the two FCA experiences (actually, even the examples from the package documentation can’t be executed anymore)</strong><a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a>. Typical pipeline is illustrated below:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#if (!requireNamespace("BiocManager", quietly = TRUE))</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co">#    install.packages("BiocManager")</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install("ArrayExpress")</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ArrayExpress)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co"># download both processed and raw datasets</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>ae_files <span class="ot">&lt;-</span> <span class="fu">getAE</span>(<span class="st">"E-MTAB-10628"</span>, <span class="at">type =</span> <span class="st">"full"</span>, <span class="at">path =</span> <span class="st">"data/Fly_Cell_Atlas_10x/"</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="do">## Build a an ExpressionSet from the raw data</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>ae_data_raw <span class="ot">&lt;-</span> <span class="fu">ae2bioc</span>(<span class="at">mageFiles =</span> ae_files)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Build a an ExpressionSet from the processed data</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>ae_data <span class="ot">&lt;-</span> <span class="fu">procset</span>(ae_files, <span class="fu">getcolproc</span>(ae_files)[<span class="dv">2</span>])</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="co"># equivalent to run ArrayExpress function</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>ae_data <span class="ot">&lt;-</span> <span class="fu">ArrayExpress</span>(<span class="st">"E-MTAB-10628"</span>, <span class="at">path =</span> <span class="st">"data/Fly_Cell_Atlas_10x/"</span>,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>                        <span class="at">save =</span> <span class="cn">TRUE</span>, <span class="at">dataCols =</span> <span class="cn">NULL</span>)</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a><span class="co"># use methods from the Biobase::ExpressionSet object to retrieve respectively </span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>pheno_data <span class="ot">&lt;-</span> <span class="fu">phenoData</span>(ae_data) <span class="co"># phenotype data</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>experiment_data <span class="ot">&lt;-</span> <span class="fu">experimentData</span>(ae_data) <span class="co"># MIAME experiment annotation (metadata)</span></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>gene_annotation <span class="ot">&lt;-</span> <span class="fu">fData</span>(ae_data) <span class="co"># gene feature annotation</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>expression_matrix <span class="ot">&lt;-</span> <span class="fu">exprs</span>(ae_data) <span class="co"># expression matrix data</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a><span class="co"># worth noted that the ExpressionSet object is well tailored for bulk micro-array </span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a><span class="co"># and RNASeq, but not really for singlecell RNASeq experiences</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively, you can download both <code>10X</code> and <code>Smartseq2</code> experiences and expression data at the count level (raw and after applying <em>TPM normalisation</em>) using the following <a href="https://www.ebi.ac.uk/gxa/sc/experiments?species=%22drosophila+melanogaster%22&amp;experimentProjects=%22Fly+Cell+Atlas%22&amp;kingdom=%22Animals%22">Single cell Expression Atlas web query</a>, then crossing out both datasets, and finally clicking on the <code>Download 2 entries</code> button feature.</p>
<p>You can perform it simultaneously for both datasets, or one dataset after the other, as illustrated in <a href="#lst-array-express-https-protocol" class="quarto-xref">Listing&nbsp;3</a> for the <code>https</code> protocol (only a subsetting of files is available through this protocol) and the <code>ftp</code> protocol, <a href="#lst-array-express-ftp-protocol" class="quarto-xref">Listing&nbsp;4</a><a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a>: <!--# write a function to automate all these download queries and checks + use req_perform_parallel to proceed all downloads in parallel for enhanced speed combined with a progress bar --></p>
<div class="cell" data-layout-align="center">
<div id="lst-array-express-https-protocol" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-array-express-https-protocol-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;3: Import ArrayExpress datasets using the https protocol
</figcaption>
<div aria-describedby="lst-array-express-https-protocol-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-array-express-https-protocol"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-array-express-https-protocol-1"><a href="#lst-array-express-https-protocol-1" aria-hidden="true" tabindex="-1"></a><span class="co"># "https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download/zip?fileType=experiment-metadata&amp;accessKey="</span></span>
<span id="lst-array-express-https-protocol-2"><a href="#lst-array-express-https-protocol-2" aria-hidden="true" tabindex="-1"></a><span class="co"># "https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download?fileType=experiment-design&amp;accessKey="</span></span>
<span id="lst-array-express-https-protocol-3"><a href="#lst-array-express-https-protocol-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download/zip?fileType=normalised&amp;accessKey=</span></span>
<span id="lst-array-express-https-protocol-4"><a href="#lst-array-express-https-protocol-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download/zip?fileType=quantification-raw&amp;accessKey=</span></span>
<span id="lst-array-express-https-protocol-5"><a href="#lst-array-express-https-protocol-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-array-express-https-protocol-6"><a href="#lst-array-express-https-protocol-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-array-express-https-protocol-7"><a href="#lst-array-express-https-protocol-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-array-express-https-protocol-8"><a href="#lst-array-express-https-protocol-8" aria-hidden="true" tabindex="-1"></a>base_url <span class="ot">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/gxa/sc/experiment/"</span></span>
<span id="lst-array-express-https-protocol-9"><a href="#lst-array-express-https-protocol-9" aria-hidden="true" tabindex="-1"></a>array_identifer <span class="ot">&lt;-</span> <span class="st">"E-MTAB-10628"</span> <span class="co"># the SmartSeq identifier</span></span>
<span id="lst-array-express-https-protocol-10"><a href="#lst-array-express-https-protocol-10" aria-hidden="true" tabindex="-1"></a><span class="co"># array_identifer &lt;- "E-MTAB-10519" # the 10X identifier</span></span>
<span id="lst-array-express-https-protocol-11"><a href="#lst-array-express-https-protocol-11" aria-hidden="true" tabindex="-1"></a><span class="co"># we took the example of the metadata zip file, containing idf and sdrf files</span></span>
<span id="lst-array-express-https-protocol-12"><a href="#lst-array-express-https-protocol-12" aria-hidden="true" tabindex="-1"></a><span class="co"># should be extended to normalised and raw expression files + experiment design</span></span>
<span id="lst-array-express-https-protocol-13"><a href="#lst-array-express-https-protocol-13" aria-hidden="true" tabindex="-1"></a>metadata_path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="st">"./data/Fly_Cell_Atlas_10x"</span>,</span>
<span id="lst-array-express-https-protocol-14"><a href="#lst-array-express-https-protocol-14" aria-hidden="true" tabindex="-1"></a>                           <span class="fu">paste0</span>(array_identifer, <span class="st">"-experiment-metadata"</span>, <span class="st">".zip"</span>))</span>
<span id="lst-array-express-https-protocol-15"><a href="#lst-array-express-https-protocol-15" aria-hidden="true" tabindex="-1"></a>fca_request <span class="ot">&lt;-</span> <span class="fu">request</span>(base_url) <span class="sc">|&gt;</span></span>
<span id="lst-array-express-https-protocol-16"><a href="#lst-array-express-https-protocol-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(array_identifer, <span class="st">"download"</span>, <span class="st">"zip"</span>) <span class="sc">|&gt;</span> </span>
<span id="lst-array-express-https-protocol-17"><a href="#lst-array-express-https-protocol-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_query</span>(<span class="at">fileType=</span><span class="st">"experiment-metadata"</span>, <span class="at">accessKey=</span><span class="st">""</span>) <span class="sc">|&gt;</span> </span>
<span id="lst-array-express-https-protocol-18"><a href="#lst-array-express-https-protocol-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_progress</span>(<span class="at">type =</span> <span class="st">"down"</span>) <span class="sc">|&gt;</span> </span>
<span id="lst-array-express-https-protocol-19"><a href="#lst-array-express-https-protocol-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>(<span class="at">path =</span> metadata_path) </span>
<span id="lst-array-express-https-protocol-20"><a href="#lst-array-express-https-protocol-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-array-express-https-protocol-21"><a href="#lst-array-express-https-protocol-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">resp_has_body</span>(fca_request) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">resp_is_error</span>(fca_request)) {</span>
<span id="lst-array-express-https-protocol-22"><a href="#lst-array-express-https-protocol-22" aria-hidden="true" tabindex="-1"></a>  zip<span class="sc">::</span><span class="fu">unzip</span>(metadata_path, </span>
<span id="lst-array-express-https-protocol-23"><a href="#lst-array-express-https-protocol-23" aria-hidden="true" tabindex="-1"></a>             <span class="at">overwrite =</span> <span class="cn">TRUE</span>, <span class="at">exdir =</span> <span class="st">"./data/Fly_Cell_Atlas_10x"</span>)</span>
<span id="lst-array-express-https-protocol-24"><a href="#lst-array-express-https-protocol-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">file.remove</span>(metadata_path)</span>
<span id="lst-array-express-https-protocol-25"><a href="#lst-array-express-https-protocol-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<div class="cell" data-layout-align="center">
<div id="lst-array-express-ftp-protocol" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-array-express-ftp-protocol-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;4: Import ArrayExpress datasets using the ftp protocol
</figcaption>
<div aria-describedby="lst-array-express-ftp-protocol-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-array-express-ftp-protocol"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-array-express-ftp-protocol-1"><a href="#lst-array-express-ftp-protocol-1" aria-hidden="true" tabindex="-1"></a>base_url <span class="ot">&lt;-</span> <span class="st">"ftp.ebi.ac.uk/pub/databases/microarray/data/atlas/sc_experiments"</span></span>
<span id="lst-array-express-ftp-protocol-2"><a href="#lst-array-express-ftp-protocol-2" aria-hidden="true" tabindex="-1"></a>array_identifier <span class="ot">&lt;-</span> <span class="st">"E-MTAB-10628"</span></span>
<span id="lst-array-express-ftp-protocol-3"><a href="#lst-array-express-ftp-protocol-3" aria-hidden="true" tabindex="-1"></a>ftp_url <span class="ot">&lt;-</span> <span class="fu">paste0</span>(base_url, <span class="st">"/"</span>, array_identifier, <span class="st">"/"</span>)</span>
<span id="lst-array-express-ftp-protocol-4"><a href="#lst-array-express-ftp-protocol-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-array-express-ftp-protocol-5"><a href="#lst-array-express-ftp-protocol-5" aria-hidden="true" tabindex="-1"></a><span class="co"># List all files in the FTP directory</span></span>
<span id="lst-array-express-ftp-protocol-6"><a href="#lst-array-express-ftp-protocol-6" aria-hidden="true" tabindex="-1"></a>ftp_filenames <span class="ot">&lt;-</span> <span class="fu">curl_fetch_memory</span>(ftp_url) <span class="sc">|&gt;</span></span>
<span id="lst-array-express-ftp-protocol-7"><a href="#lst-array-express-ftp-protocol-7" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">pluck</span>(<span class="st">"content"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-array-express-ftp-protocol-8"><a href="#lst-array-express-ftp-protocol-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rawToChar</span>() <span class="sc">|&gt;</span></span>
<span id="lst-array-express-ftp-protocol-9"><a href="#lst-array-express-ftp-protocol-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Split by rows (newline character)</span></span>
<span id="lst-array-express-ftp-protocol-10"><a href="#lst-array-express-ftp-protocol-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_split</span>( <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\r\n</span><span class="st">"</span>, <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="lst-array-express-ftp-protocol-11"><a href="#lst-array-express-ftp-protocol-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_subset</span>(<span class="st">"[^ ]"</span>) <span class="sc">|&gt;</span>  <span class="co"># remove empty lines |&gt;</span></span>
<span id="lst-array-express-ftp-protocol-12"><a href="#lst-array-express-ftp-protocol-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># matches one or more characters that are not whitespace, to the end of the string.</span></span>
<span id="lst-array-express-ftp-protocol-13"><a href="#lst-array-express-ftp-protocol-13" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_extract</span>( <span class="at">pattern =</span> <span class="st">"[^</span><span class="sc">\\</span><span class="st">s]+$"</span>)</span>
<span id="lst-array-express-ftp-protocol-14"><a href="#lst-array-express-ftp-protocol-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-array-express-ftp-protocol-15"><a href="#lst-array-express-ftp-protocol-15" aria-hidden="true" tabindex="-1"></a>idf_index <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="at">pattern =</span> <span class="st">"idf"</span>, ftp_URLs)</span>
<span id="lst-array-express-ftp-protocol-16"><a href="#lst-array-express-ftp-protocol-16" aria-hidden="true" tabindex="-1"></a>ftp_URLs <span class="ot">&lt;-</span> <span class="fu">paste0</span>(ftp_url, ftp_filenames)</span>
<span id="lst-array-express-ftp-protocol-17"><a href="#lst-array-express-ftp-protocol-17" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(ftp_URLs[idf_index],</span>
<span id="lst-array-express-ftp-protocol-18"><a href="#lst-array-express-ftp-protocol-18" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> ftp_filenames[idf_index],</span>
<span id="lst-array-express-ftp-protocol-19"><a href="#lst-array-express-ftp-protocol-19" aria-hidden="true" tabindex="-1"></a>              <span class="at">method =</span> <span class="st">"auto"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="ArrayExpress datasets">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
ArrayExpress datasets
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>ArrayExpress datasets are traditionally organized into the following key file types, ordered by increasing level of detail:</p>
<ul>
<li><p><strong>Investigation Description Format (IDF)</strong> provides a high-level overview of the experiment.</p>
<ul>
<li><strong>Title and Description</strong>: Summary of the experiment’s purpose.</li>
<li><strong>Design</strong> and <strong>Protocols</strong>: Experimental design and factors studied (e.g., treatments, time points).</li>
<li><strong>Contacts</strong>: Information about the authors or submitters.</li>
</ul></li>
<li><p><strong>Sample and Data Relationship Format (SDRF)</strong>: the phenotype data, detailing the mapping between <strong>Sample Identifiers</strong>, <strong>Factor Values</strong> and <strong>Data File Names</strong></p></li>
<li><p><strong>Array Design Format (ADF)</strong> used to provide the design of microarray platforms used in the experiment. Equivalent to <strong>featureData</strong>, setting up the mapping between the genes sampled and their modern and standard equivalents.</p></li>
<li><p>Expression matrices, as:</p>
<ul>
<li><strong>Raw Data Files</strong> (from FASTQ (for sequencing data) to CEL (for Affymetrix arrays))</li>
<li><strong>Processed Data Files</strong> (usually counts, after normalization or transformation)</li>
</ul></li>
</ul>
</div>
</div>
</div>
</section>
</section>
<section id="retrieve-transcriptomic-annotation" class="level2">
<h2 class="anchored" data-anchor-id="retrieve-transcriptomic-annotation">Retrieve transcriptomic annotation</h2>
<section id="sec-modencode-expression" class="level3">
<h3 class="anchored" data-anchor-id="sec-modencode-expression">The ModENCODE expression matrix</h3>
<p>The current RPKM database storing all bulk RNASeq experiences on the comprehensive FlyBase data collection can be retrieved <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#RNA-Seq_RPKM_values_.28gene_rpkm_report_fb_.2A.tsv.gz.29">here</a>.</p>
<p>Alternatively, we can explore the FTP repository, sub-setting the files’ selection to those mentioning <code>gene_rpkm</code> explicitly in their name<a href="#fn10" class="footnote-ref" id="fnref10" role="doc-noteref"><sup>10</sup></a> <!-- Develop a custom R function to parse files retrieved from a FTP repository --></p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># FTP URL for the repository</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>ftp_url <span class="ot">&lt;-</span> <span class="st">"ftp://ftp.flybase.net/releases/current/precomputed_files/genes/"</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># List all files in the FTP directory</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>ftp_files <span class="ot">&lt;-</span> <span class="fu">curl_fetch_memory</span>(ftp_url) <span class="sc">|&gt;</span> </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  purrr<span class="sc">::</span><span class="fu">pluck</span>(<span class="st">"content"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rawToChar</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Split by rows (newline character)</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_split</span>( <span class="at">pattern =</span> <span class="st">"</span><span class="sc">\r\n</span><span class="st">"</span>, <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span> </span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_subset</span>(<span class="st">"[^ ]"</span>) <span class="co"># remove empty lines</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Split each row by columns (tab character) and convert to a data.frame</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>ftp_files <span class="ot">&lt;-</span> <span class="fu">as_tibble</span>(<span class="fu">do.call</span>(rbind, </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>                               <span class="fu">str_split</span>(ftp_files, <span class="at">pattern =</span> <span class="st">"[[:blank:]]{2,}"</span>)))</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>ftp_files <span class="ot">&lt;-</span> ftp_files <span class="sc">|&gt;</span> </span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">separate_wider_delim</span>(<span class="at">cols =</span> <span class="st">"V4"</span>,</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>                              <span class="at">delim =</span> <span class="st">" "</span>, </span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>                              <span class="at">too_many =</span> <span class="fu">c</span>(<span class="st">"drop"</span>), <span class="co"># silently remove file renaming</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>                              <span class="at">names =</span> <span class="fu">c</span>(<span class="st">"size"</span>, <span class="st">"month"</span>, <span class="st">"day"</span>, <span class="st">"hour"</span>, <span class="st">"filename"</span>))</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>rnaseq_ftp_files <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_subset</span>(ftp_files<span class="sc">$</span>filename, </span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">pattern =</span> <span class="st">"gene_rpkm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In that case, only 2 store gene expression data, namely gene_rpkm_matrix_fb_2024_05.tsv.gz, gene_rpkm_report_fb_2024_05.tsv.gz:</p>
<ul>
<li>gene_rpkm_matrix_fb_2024_05.tsv.gz: <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#RNA-Seq_RPKM_values_matrix_.28gene_rpkm_matrix_fb_.2A.tsv.gz.29">Table description</a>. <strong>Note that <code>ModENCODE</code> data is provided with RPKM pre-processed RNA-Seq expression values, while <code>FlyAtlas2</code> data has been normalised to FPKM units.</strong></li>
<li>gene_rpkm_report_fb_2024_05.tsv.gz: <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#RNA-Seq_RPKM_values_.28gene_rpkm_report_fb_.2A.tsv.gz.29">Table description</a></li>
</ul>
</section>
<section id="the-flyatlas2-expression-matrix" class="level3">
<h3 class="anchored" data-anchor-id="the-flyatlas2-expression-matrix">The FlyAtlas2 expression matrix</h3>
</section>
<section id="the-fly-cell-atlas-expression-matrix" class="level3">
<h3 class="anchored" data-anchor-id="the-fly-cell-atlas-expression-matrix">The Fly Cell Atlas expression matrix</h3>
<p>Depending on the level of resolution, the new <code>Biostudies</code> website, aiming at replacing the <code>ArrayExpress</code> web server, provides two APIs:</p>
<ol type="1">
<li><p>To download files at the expression level (counts), both normalised and raw counts, we already detailed a programmatic process in <a href="#sec-fca-pheno-data" class="quarto-xref">Section&nbsp;1.2.4</a>. Note that in addition to the protocol detailed in previous section, you may alternatively use <code>FTP</code>, <code>curl</code>, <code>Aspera</code> or <code>Globus</code> software. However, these tools usually require to know in advance all file locations and paths to be downloaded. All these tools are further detailed in <a href="https://www.ebi.ac.uk/biostudies/help#">Biostudies documention</a>.</p></li>
<li><p>The ENA Portal API (see <a href="#lst-ena-ebi-api" class="quarto-xref">Listing&nbsp;5</a> for programmatic access) can be used to retrieve raw files, at the sequence level (in other words, the bio-informatician job, with possibility to retrieve both FASTQ and BAM files, the latter corresponding to sequences already aligned against a reference genome).</p></li>
</ol>
<ul>
<li>It requires to retrieve the corresponding ENA or Project accession number (usually, the <code>PRJ...</code> is the main identifier, while the <code>ERP...</code> is the second, optional identifier). <code>E-MTAB-10628</code> is paired with <code>PRJEB45993</code> project (ENA identifier: <code>ERP130174</code>) and, while <code>E-MTAB-10519</code> is paired with <code>PRJEB45570</code> (ENA identifier: <code>ERP129698</code>).</li>
<li>More documentation details can be found <a href="https://ena-docs.readthedocs.io/en/latest/retrieval/programmatic-access.html">here</a>, and an interactive swagger UI is reported <a href="https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html">here</a>.</li>
<li>Without this mapping information, you can browse it on the <code>Biostudies</code> website (report to <a href="#fig-ERP-MAGE" class="quarto-xref">Figure&nbsp;1</a> for a snapshot), or programmatically by either parsing the <code>idf</code> file or retrieving the json file associated with the Biostudies online accession (see <a href="#lst-mtab-ena-mapping" class="quarto-xref">Listing&nbsp;6</a>).</li>
</ul>
<div id="fig-ERP-MAGE" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ERP-MAGE-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="img/E-MTAB-10519-ERR-correspondance.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ERP-MAGE-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: ERP and ExpressionSet MAGE mapping
</figcaption>
</figure>
</div>
<div class="cell" data-layout-align="center">
<div id="lst-ena-ebi-api" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-ena-ebi-api-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;5: The ENA Portal API
</figcaption>
<div aria-describedby="lst-ena-ebi-api-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-ena-ebi-api"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-ena-ebi-api-1"><a href="#lst-ena-ebi-api-1" aria-hidden="true" tabindex="-1"></a><span class="co"># interactive swagger ui https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html#/Search%20%26%20Discovery/getResults</span></span>
<span id="lst-ena-ebi-api-2"><a href="#lst-ena-ebi-api-2" aria-hidden="true" tabindex="-1"></a>url_base <span class="ot">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/ena/portal/api"</span></span>
<span id="lst-ena-ebi-api-3"><a href="#lst-ena-ebi-api-3" aria-hidden="true" tabindex="-1"></a><span class="co"># all available result types for a given study:</span></span>
<span id="lst-ena-ebi-api-4"><a href="#lst-ena-ebi-api-4" aria-hidden="true" tabindex="-1"></a>ena_avalaible_datasets <span class="ot">&lt;-</span> <span class="fu">request</span>(url_base) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-5"><a href="#lst-ena-ebi-api-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(<span class="st">"results"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-6"><a href="#lst-ena-ebi-api-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_query</span>(<span class="at">format=</span><span class="st">"tsv"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-7"><a href="#lst-ena-ebi-api-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>() <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-8"><a href="#lst-ena-ebi-api-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resp_body_string</span>() <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-9"><a href="#lst-ena-ebi-api-9" aria-hidden="true" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="lst-ena-ebi-api-10"><a href="#lst-ena-ebi-api-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-ena-ebi-api-11"><a href="#lst-ena-ebi-api-11" aria-hidden="true" tabindex="-1"></a><span class="co"># of interest for raw files, the `read_run` is certainly the most complete</span></span>
<span id="lst-ena-ebi-api-12"><a href="#lst-ena-ebi-api-12" aria-hidden="true" tabindex="-1"></a><span class="co"># request for retrieving all potential fields for the read_run dataset</span></span>
<span id="lst-ena-ebi-api-13"><a href="#lst-ena-ebi-api-13" aria-hidden="true" tabindex="-1"></a><span class="co"># interactive: https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html#/Search%20%26%20Discovery/getReturnFields</span></span>
<span id="lst-ena-ebi-api-14"><a href="#lst-ena-ebi-api-14" aria-hidden="true" tabindex="-1"></a>read_run_fields <span class="ot">&lt;-</span> <span class="fu">request</span>(url_base) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-15"><a href="#lst-ena-ebi-api-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(<span class="st">"returnFields"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-16"><a href="#lst-ena-ebi-api-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_query</span>(<span class="at">dataPortal=</span><span class="st">"ena"</span>, <span class="at">result=</span><span class="st">"read_run"</span>, <span class="at">format=</span><span class="st">"tsv"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-17"><a href="#lst-ena-ebi-api-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>() <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-18"><a href="#lst-ena-ebi-api-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resp_body_string</span>() <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-19"><a href="#lst-ena-ebi-api-19" aria-hidden="true" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="lst-ena-ebi-api-20"><a href="#lst-ena-ebi-api-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-ena-ebi-api-21"><a href="#lst-ena-ebi-api-21" aria-hidden="true" tabindex="-1"></a><span class="co"># and finally, for a given project, the direct FTP links to fastq or BAM files</span></span>
<span id="lst-ena-ebi-api-22"><a href="#lst-ena-ebi-api-22" aria-hidden="true" tabindex="-1"></a><span class="co"># interactive: https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html#/Search%20%26%20Discovery/fileReport</span></span>
<span id="lst-ena-ebi-api-23"><a href="#lst-ena-ebi-api-23" aria-hidden="true" tabindex="-1"></a>PRJEB45993_BAM_FASTQ_URLs <span class="ot">&lt;-</span> <span class="fu">request</span>(url_base) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-24"><a href="#lst-ena-ebi-api-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(<span class="st">"filereport"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-25"><a href="#lst-ena-ebi-api-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_query</span>(<span class="at">accession=</span><span class="st">"PRJEB45993"</span>,</span>
<span id="lst-ena-ebi-api-26"><a href="#lst-ena-ebi-api-26" aria-hidden="true" tabindex="-1"></a>                <span class="at">result=</span><span class="st">"read_run"</span>,</span>
<span id="lst-ena-ebi-api-27"><a href="#lst-ena-ebi-api-27" aria-hidden="true" tabindex="-1"></a>                <span class="at">fields=</span><span class="fu">c</span>(<span class="st">"study_accession"</span>, <span class="st">"experiment_accession"</span>,<span class="st">"run_accession"</span>,</span>
<span id="lst-ena-ebi-api-28"><a href="#lst-ena-ebi-api-28" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"fastq_ftp"</span>, <span class="st">"fastq_md5"</span>,</span>
<span id="lst-ena-ebi-api-29"><a href="#lst-ena-ebi-api-29" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"bam_ftp"</span>,<span class="st">" bam_md5"</span>),</span>
<span id="lst-ena-ebi-api-30"><a href="#lst-ena-ebi-api-30" aria-hidden="true" tabindex="-1"></a>                <span class="at">format=</span><span class="st">"tsv"</span>,</span>
<span id="lst-ena-ebi-api-31"><a href="#lst-ena-ebi-api-31" aria-hidden="true" tabindex="-1"></a>                <span class="at">download =</span> <span class="cn">TRUE</span>, </span>
<span id="lst-ena-ebi-api-32"><a href="#lst-ena-ebi-api-32" aria-hidden="true" tabindex="-1"></a>                <span class="at">.multi =</span> <span class="st">"comma"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-33"><a href="#lst-ena-ebi-api-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>() </span>
<span id="lst-ena-ebi-api-34"><a href="#lst-ena-ebi-api-34" aria-hidden="true" tabindex="-1"></a><span class="co"># for whatever reason, piping all these operations lead to time out. </span></span>
<span id="lst-ena-ebi-api-35"><a href="#lst-ena-ebi-api-35" aria-hidden="true" tabindex="-1"></a>read_run_fields <span class="ot">&lt;-</span> read_run_fields <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-36"><a href="#lst-ena-ebi-api-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resp_body_string</span>() <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-37"><a href="#lst-ena-ebi-api-37" aria-hidden="true" tabindex="-1"></a>  readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="lst-ena-ebi-api-38"><a href="#lst-ena-ebi-api-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-ena-ebi-api-39"><a href="#lst-ena-ebi-api-39" aria-hidden="true" tabindex="-1"></a><span class="co"># number of rows:</span></span>
<span id="lst-ena-ebi-api-40"><a href="#lst-ena-ebi-api-40" aria-hidden="true" tabindex="-1"></a><span class="co"># https://www.ebi.ac.uk/ena/portal/api/filereportcount?result=read_run&amp;accession=PRJEB45993&amp;format=tsv</span></span>
<span id="lst-ena-ebi-api-41"><a href="#lst-ena-ebi-api-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-ena-ebi-api-42"><a href="#lst-ena-ebi-api-42" aria-hidden="true" tabindex="-1"></a>read_run_rows <span class="ot">&lt;-</span> <span class="fu">request</span>(url_base) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-43"><a href="#lst-ena-ebi-api-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(<span class="st">"filereportcount"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-44"><a href="#lst-ena-ebi-api-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_query</span>(<span class="at">accession=</span><span class="st">"PRJEB45993"</span>,</span>
<span id="lst-ena-ebi-api-45"><a href="#lst-ena-ebi-api-45" aria-hidden="true" tabindex="-1"></a>                <span class="at">result=</span><span class="st">"read_run"</span>,</span>
<span id="lst-ena-ebi-api-46"><a href="#lst-ena-ebi-api-46" aria-hidden="true" tabindex="-1"></a>                <span class="at">format=</span><span class="st">"json"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-ena-ebi-api-47"><a href="#lst-ena-ebi-api-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>() <span class="sc">|&gt;</span> </span>
<span id="lst-ena-ebi-api-48"><a href="#lst-ena-ebi-api-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resp_body_json</span>() <span class="sc">|&gt;</span> </span>
<span id="lst-ena-ebi-api-49"><a href="#lst-ena-ebi-api-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pluck</span>(<span class="st">"count"</span>) <span class="sc">|&gt;</span> </span>
<span id="lst-ena-ebi-api-50"><a href="#lst-ena-ebi-api-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.integer</span>()</span>
<span id="lst-ena-ebi-api-51"><a href="#lst-ena-ebi-api-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="lst-ena-ebi-api-52"><a href="#lst-ena-ebi-api-52" aria-hidden="true" tabindex="-1"></a>testthat<span class="sc">::</span><span class="fu">expect_equal</span>(<span class="fu">nrow</span>(read_run_fields), </span>
<span id="lst-ena-ebi-api-53"><a href="#lst-ena-ebi-api-53" aria-hidden="true" tabindex="-1"></a>                       read_run_rows)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<div class="cell" data-layout-align="center">
<div id="lst-mtab-ena-mapping" class="r cell-code listing quarto-float quarto-figure quarto-figure-left anchored">
<figure class="quarto-float quarto-float-lst figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-lst" id="lst-mtab-ena-mapping-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Listing&nbsp;6: The ENA Portal API
</figcaption>
<div aria-describedby="lst-mtab-ena-mapping-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="sourceCode cell-code" id="lst-mtab-ena-mapping"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="lst-mtab-ena-mapping-1"><a href="#lst-mtab-ena-mapping-1" aria-hidden="true" tabindex="-1"></a>url_base <span class="ot">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/biostudies"</span></span>
<span id="lst-mtab-ena-mapping-2"><a href="#lst-mtab-ena-mapping-2" aria-hidden="true" tabindex="-1"></a>array_identifier <span class="ot">&lt;-</span> <span class="st">"E-MTAB-10519"</span></span>
<span id="lst-mtab-ena-mapping-3"><a href="#lst-mtab-ena-mapping-3" aria-hidden="true" tabindex="-1"></a>ERP_symbol <span class="ot">&lt;-</span> <span class="fu">request</span>(url_base) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-4"><a href="#lst-mtab-ena-mapping-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(<span class="st">"files"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-5"><a href="#lst-mtab-ena-mapping-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(array_identifier) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-6"><a href="#lst-mtab-ena-mapping-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(<span class="fu">paste0</span>(array_identifier, <span class="st">".json"</span>)) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-7"><a href="#lst-mtab-ena-mapping-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>() <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-8"><a href="#lst-mtab-ena-mapping-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resp_body_json</span>(<span class="at">simplifyVector =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-9"><a href="#lst-mtab-ena-mapping-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pluck</span>(<span class="st">"section"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-10"><a href="#lst-mtab-ena-mapping-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pluck</span>(<span class="st">"links"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-11"><a href="#lst-mtab-ena-mapping-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pluck</span>(<span class="dv">1</span>) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-12"><a href="#lst-mtab-ena-mapping-12" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">pull</span>(<span class="st">"url"</span>) <span class="sc">|&gt;</span></span>
<span id="lst-mtab-ena-mapping-13"><a href="#lst-mtab-ena-mapping-13" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_subset</span>(<span class="st">"^ERP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</figure>
</div>
</div>
<p>Of note, without REST API programmatic access, the <a href="https://www.bioconductor.org/packages//2.10/bioc/vignettes/SRAdb/inst/doc/SRAdb.pdf"><strong>BioConductor SRAdb package</strong></a> provides programmatic access to metadata from the Sequence Read Archive (SRA) and ENA (European Nucleotide Archive) from R<a href="#fn11" class="footnote-ref" id="fnref11" role="doc-noteref"><sup>11</sup></a>.</p>
<p>However, it relies on a strong SQLite dependency, has not been maintained for more than two years, and requires prior installation of the SQL metadata database, implying strong storage and memory-resource. Besides, search of the database has been streamlined by the development of the <a href="https://www.ebi.ac.uk/ena/browser/api/swagger-ui/index.html">ENA Browser API</a>.</p>
<p>Other methods for non-programmers are reported in <a href="https://ena-docs.readthedocs.io/en/latest/retrieval/file-download.html#downloading-files">Download Files ENA Documentation</a>, with popular tools like <code>ENA Browser</code>, <code>ENA FTP Downloader GUI tool</code>, <code>Globus</code> or <code>Aspera</code>. While practical for downloading thousands of files, with dedicated facilities to manage parallel downloading or error handling, they require custom and often admin rights’ installation, depending on third-party software. Besides, they’re not really practical for enforcing FAIR and reproducibility guidelines.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="MD5 role">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
MD5 role
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You may have noticed in <a href="#lst-ena-ebi-api" class="quarto-xref">Listing&nbsp;5</a> that we have collected both original raw files and their MD5 peers. Indeed, the <strong>MD5 file</strong> stores a 32-character hexadecimal string generated that guarantees data integrity (check for file’s alterations or corruptions, partial downloading, …). Any mismatch in checksums indicates that the file has been corrupted. The easiest way to verify file integrity is certainly by looping over each pair of original file - MD5 file, using the <code>digest</code> package:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(digest)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the file path and expected MD5 checksum</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>file_path <span class="ot">&lt;-</span> <span class="st">"path/to/your/file.txt"</span>  <span class="co"># Replace with your file's path</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>expected_md5 <span class="ot">&lt;-</span> <span class="st">"d41d8cd98f00b204e9800998ecf8427e"</span>  <span class="co"># Replace with the provided checksum</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the MD5 checksum of the file</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>calculated_md5 <span class="ot">&lt;-</span> <span class="fu">digest</span>(<span class="at">file =</span> file_path, <span class="at">algo =</span> <span class="st">"md5"</span>, <span class="at">serialize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co"># testthat::expect_equal(calculated_md5, expected_md5)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="optionnal-ftp-access-on-biostudies-and-not-ena" class="level4">
<h4 class="anchored" data-anchor-id="optionnal-ftp-access-on-biostudies-and-not-ena">(Optionnal) FTP access on Biostudies (and not ENA)</h4>
<p>We detailed in the following code snippet:</p>
<ol type="1">
<li><p>the <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/help#programmatic">Biostudies API</a> to retrieve the most comprehensive metadata file for a given study, the latter listing also differences with the previous ArrayExpress web server.</p></li>
<li><p>A <a href="https://www.rdocumentation.org/packages/RCurl/versions/1.98-1.16/topics/getURL"><code>RCurl</code></a> simplified code to list and download all files along with their properties (such as size or access rights). Yet, the Biostudies repository is usually much less comprehensive and organised as the ENA/SRA web server, while exhibiting the following odd path composition: <code>ftp.sra.ebi.ac.uk/vol1/&lt;file-type&gt;:err|fastq|run/&lt;accession-prefix&gt;/&lt;00-last-digit-of-full-accession&gt;/&lt;full-accession&gt;/</code>, described thoroughly in <a href="https://ena-docs.readthedocs.io/en/latest/retrieval/file-download/sra-ftp-structure.html">SRA FTP Structure documentation</a>, and preventing downloading the whole study without access to metadata and file localisations.</p></li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1) not that relevant, too many fields to process, without clear structure</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>url_base <span class="ot">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/biostudies/api/v1/studies"</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>array_identifier <span class="ot">&lt;-</span> <span class="st">"E-MTAB-10519"</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>ena_detailled_information <span class="ot">&lt;-</span> <span class="fu">request</span>(url_base) <span class="sc">|&gt;</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_url_path_append</span>(array_identifier) <span class="sc">|&gt;</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">req_perform</span>() <span class="sc">|&gt;</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resp_body_json</span>()</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 2) easiest process for listing all files in a FTP repo using RCurl</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>base_url <span class="ot">&lt;-</span> <span class="st">'ftp.ebi.ac.uk/biostudies'</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>study_type <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(array_identifier, <span class="st">"^[[[:alpha:]]-]+"</span>)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>study_suffix <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(array_identifier, <span class="st">"[[:digit:]]{3}$"</span>)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the Files/ if you just wish to access general metadata</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>biostudies_ftp_url <span class="ot">&lt;-</span> <span class="fu">paste</span>(base_url, <span class="st">"fire"</span>,</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>                            study_type, study_suffix,</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>                            array_identifier, <span class="st">"Files/"</span>,</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>                            <span class="at">sep =</span> <span class="st">"/"</span>)</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a><span class="co"># ftp.use.epsv ensures that the request is not rejected by the FTP web server</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a><span class="co"># dirlistonly ensures that we only retrieve filenames from the FTP repository</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>ftp_biostudies <span class="ot">&lt;-</span> RCurl<span class="sc">::</span><span class="fu">getURL</span>(biostudies_ftp_url,</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>                           <span class="at">ftp.use.epsv =</span> <span class="cn">FALSE</span>, <span class="at">dirlistonly =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_split</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\r\n</span><span class="st">"</span>, <span class="at">simplify =</span> <span class="cn">TRUE</span>) <span class="sc">|&gt;</span></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.vector</span>() <span class="sc">|&gt;</span></span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">str_subset</span>(<span class="st">"[^[[:blank:]]*]"</span>) <span class="co"># remove empty chains, only composed of white spaces</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a><span class="co"># define the connection port</span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>con <span class="ot">&lt;-</span>  <span class="fu">getCurlHandle</span>(<span class="at">ftp.use.epsv =</span> <span class="cn">FALSE</span>)</span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a><span class="co"># download all file listed in the FTP repository</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>contents <span class="ot">&lt;-</span>  <span class="fu">sapply</span>(filenames, getURL, <span class="at">curl =</span> con)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="optionnal-sra-and-ena-organisational-structure" class="level4">
<h4 class="anchored" data-anchor-id="optionnal-sra-and-ena-organisational-structure">(Optionnal) SRA and ENA organisational structure</h4>
<div class="callout callout-style-default callout-note callout-titled" title="ENA hierarchical structure">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
ENA hierarchical structure
</div>
</div>
<div class="callout-body-container callout-body">
<p>The hierarchical structure in the <strong>SRA (Sequence Read Archive)</strong> or <strong>ENA (European Nucleotide Archive)</strong> follows the following order of granularity:</p>
<p>1.<strong>ERP (or SRP)</strong> represents a <strong>study or project</strong>, describing the overarching goal of the sequencing experiment and providing metadata about the research context.</p>
<ol start="2" type="1">
<li><p><strong>ERS (or SRS)</strong> represents an <strong>individual biological sample</strong> within a study.</p></li>
<li><p><strong>ERX (or SRX)</strong> represents an <strong>experiment</strong>, describing the methodology applied to a sample, such as library preparation, sequencing platform and sequencing strategy (e.g., RNA-seq, WGS).</p></li>
<li><p>The <strong>ERR (or SRR)</strong> represents a specific <strong>run</strong>, which is the unit of raw sequencing data generated from an experiment. Each ERR corresponds to a FASTQ file or set of FASTQ files containing sequencing reads.</p></li>
</ol>
<p>Note that complex mappings relate these objects to each other:</p>
<ul>
<li>Multiple Experiments can be grouped together to form a Study.</li>
<li>A single Sample can be used by more than one Experiment, so there is a many-to-many relationship between Samples and Experiments.</li>
<li>One Experiment can store multiple Runs.</li>
<li>Only the Run level has a one-to-one relationship or Paired-End, assuming the FASTQ format was used, relationship with the original files.</li>
</ul>
</div>
</div>
<p>More documentation on the - Main facilities delivered by the ENA web server, including programmatic access: <a href="https://pubmed.ncbi.nlm.nih.gov/18978013/">Petabyte-scale innovations at the European Nucleotide Archive</a>, from <span class="citation" data-cites="cochrane2009nar">(<a href="#ref-cochrane2009nar" role="doc-biblioref"><strong>cochrane2009nar?</strong></a>)</span>. - Some infographics for detailing the interconnections between the most prominent gene repositories are reported in Fig. <a href="#fig-sra-structure" class="quarto-xref">Figure&nbsp;2</a>.</p>
<div id="fig-sra-structure" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-sra-structure-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/sra_databases.png" class="img-fluid figure-img"></p>
<figcaption><a href="https://www.ccdatalab.org/blog/gene-expression-repositories-explained">Gene Expression Repositories Explained</a></figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/sra_pipeline.png" class="img-fluid figure-img"></p>
<figcaption>Pipeline to pair raw data from ENA/SRA, metadata from Biosamples and reference genoms from Ensembl and RefSeq</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row">
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/sra_structure.png" class="img-fluid figure-img"></p>
<figcaption>SRA structure</figcaption>
</figure>
</div>
</div>
<div class="quarto-layout-cell" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/ena_structure.jpg" class="img-fluid figure-img"></p>
<figcaption>The relational data model for ENA reads: a study is associated with samples that are themselves related to runs. Besides, each run is associated with an experiment identifier describing which technologu has been used to perform it. Underlying this data model is an API that provides abstraction from the nature of the data file system, returning read data upon request based on read identifiers.</figcaption>
</figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-sra-structure-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: ENA and SRA structures and layout.
</figcaption>
</figure>
</div>
</section>
</section>
</section>
</section>



<div id="quarto-appendix" class="default"><section id="programmatic-access-to-flybase" class="level1 appendix"><h2 class="anchored quarto-appendix-heading">Programmatic access to FlyBase</h2><div class="quarto-appendix-contents">

<p>To retrieve datasets stored in FlyBase, such as sex, tissue, or lab information, in an automated and programmatic manner, you can use of these three approaches, ordered by increasing complexity but scalability:</p>
<hr>
<section id="precomputed-metadata-files" class="level2">
<h2 class="anchored" data-anchor-id="precomputed-metadata-files"><strong>Precomputed Metadata Files</strong>:</h2>
<p>They are accessible using the <a href="https://flybase.org/downloads/bulkdata">FlyBase FTP</a>: - <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview">Wiki Documentation of the Databases structure and Overview</a></p>
<div class="callout callout-style-default callout-note callout-titled" title="Only for Linux users: wget access">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Only for Linux users: wget access
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li><strong>Download the RPKM file</strong>:</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> http://ftp.flybase.org/releases/FB2024_05/precomputed_files/genes/gene_rpkm_report_fb_<span class="op">&lt;</span>version<span class="op">&gt;</span>.tsv.gz </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>where <code>&lt;version&gt;</code> is a <em>placeholder</em> referring to a FlyBase release. Alternatively, use:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> http://ftp.flybase.org/releases/FB2024_05/precomputed_files/genes/gene_rpkm_<span class="pp">*</span>.tsv.gz </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>where <code>*</code> is a regex expression standing for any symbol (in that case, it should download two datasets).</p>
<ol start="2" type="1">
<li><strong>Extract and view the file</strong>:</li>
</ol>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gunzip</span> gene_rpkm_report_fb_2024_05.tsv.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol start="3" type="1">
<li><strong>Analyze in R</strong>:</li>
</ol>
<div class="sourceCode" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>RPKM_file <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(<span class="st">"gene_rpkm_report_fb_2024_05.tsv"</span>, </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(RPKM_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Alternatively, we propose the following portable R script to replicate the <code>wget</code> and <code>gunzip</code> functionalities.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the URL and output file paths</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"http://ftp.flybase.org/releases/FB2024_05/precomputed_files/genes/gene_rpkm_report_fb_2024_05.tsv.gz"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>output_file <span class="ot">&lt;-</span> <span class="st">"data/RNASeq/gene_rpkm_report_fb_2024_05.tsv.gz"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>output_unzipped <span class="ot">&lt;-</span> <span class="st">"data/RNASeq/gene_rpkm_report_fb_2024_05.tsv"</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Download the file</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>response <span class="ot">&lt;-</span> <span class="fu">GET</span>(url, <span class="co"># `write_disk()` specifies path location of the download file.</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>                <span class="fu">write_disk</span>(output_file, <span class="at">overwrite =</span> <span class="cn">TRUE</span>))</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if download was successful</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (response<span class="sc">$</span>status_code <span class="sc">==</span> <span class="dv">200</span>) {</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"File downloaded successfully to:"</span>, output_file, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stop</span>(<span class="st">"Download failed with status code:"</span>, response<span class="sc">$</span>status_code)</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Decompress the file</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>R.utils<span class="sc">::</span><span class="fu">gunzip</span>(output_file, output_unzipped, <span class="at">remove =</span> <span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="using-the-flybase-rest-api-lightweight-approach" class="level2">
<h2 class="anchored" data-anchor-id="using-the-flybase-rest-api-lightweight-approach"><strong>Using the FlyBase REST API lightweight approach</strong>:</h2>
<p>The REST API allows for dynamic metadata retrieval via programmatic queries. In particular, the <code>GET</code> request is used to retrieve data from a server. It is however restricted by URL length. Here’s a linux example below, using the <code>curl</code> command, and using one parameter composed of several IDs separated by commas:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-X</span> GET <span class="st">"https://api.flybase.org/api/v1.0/hitlist/fetch/FBlc0000215,FBlc0000216,FBlc0000217"</span> <span class="at">-H</span> <span class="st">"Accept: application/json"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Report to <span class="citation" data-cites="modencode-metadata">(<a href="#ref-modencode-metadata" role="doc-biblioref"><strong>modencode-metadata?</strong></a>)</span> for a R high-user interface, using the complementary between the <code>jsonlite</code> and <code>httr2</code> packages.</p>
<hr>
</section>
<section id="using-the-flybase-chado-postgresql-database" class="level2">
<h2 class="anchored" data-anchor-id="using-the-flybase-chado-postgresql-database"><strong>Using the FlyBase Chado PostgreSQL Database</strong></h2>
<p>That’s definitely the most challenging, but also the safest and most scalable approach. In brief, you import the whole database and its relational scheme to your local computer or server.</p>
<p>The <em>Chado database</em> is a wrapper of the <code>PostgreSQL</code> DBMS, with custom functions to parse biological samples.</p>
<div id="nte-chado-basic" class="callout callout-style-default callout-note callout-titled" title="Use the Chado Database">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note&nbsp;1: Use the Chado Database
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="1">
<li>Connect to the FlyBase public instance<a href="#fn12" class="footnote-ref" id="fnref12" role="doc-noteref"><sup>12</sup></a>.</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">psql</span> <span class="at">-h</span> chado.flybase.org <span class="at">-U</span> flybase flybase</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Query metadata for datasets:</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode sql code-with-copy"><code class="sourceCode sql"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="kw">SELECT</span> <span class="op">*</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="kw">FROM</span> dataset</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="kw">WHERE</span> dataset_id <span class="op">=</span> <span class="st">'FB2024_05'</span>;</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="1">
<li>Export results for further analysis:</li>
</ol>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">\COPY</span> <span class="er">(</span><span class="ex">SELECT</span> <span class="pp">*</span> FROM dataset WHERE dataset_id = <span class="st">'FB2024_05'</span><span class="kw">)</span> <span class="ex">TO</span> <span class="st">'metadata.csv'</span> CSV HEADER<span class="kw">;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout-info" title="Alternative use of the Chado Database">
<p>In <a href="#nte-chado-basic" class="quarto-xref">Note&nbsp;1</a>, we described the <em>Direct Chado Query access</em>. However, this public and read-only access, directly exploitable with an installed SQL client, is suitable only for occasional access, and with lots of users requesting the database, can induce a lot of latency.</p>
<p>The alternative approach consists of first <strong>downloading and Hosting the Chado Database Locally</strong> :</p>
<ul>
<li><strong>Step 1:</strong> Download the PostgreSQL dump files from the FlyBase FTP site:<br>
<a href="ftp://ftp.flybase.org/releases/current/psql/">FlyBase PostgreSQL Dumps</a>.</li>
<li><strong>Step 2:</strong> Create a new PostgreSQL database locally:</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">createdb</span> <span class="at">-E</span> UTF-8 my_flybase</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><strong>Step 3:</strong> Load the downloaded data into your local database:</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> FB<span class="pp">*</span>.sql.gz.<span class="pp">*</span> <span class="kw">|</span> <span class="fu">gunzip</span> <span class="kw">|</span> <span class="ex">psql</span> my_flybase</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><strong>Step 4:</strong> Optimize the database<a href="#fn13" class="footnote-ref" id="fnref13" role="doc-noteref"><sup>13</sup></a> for use:</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">vacuumdb</span> <span class="at">-f</span> <span class="at">-z</span> <span class="at">-v</span> my_flybase</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><strong>Chado Schema</strong>: Learn the database schema structure to optimize queries with the <a href="https://flybase.github.io/docs/chado/functions">Chado Documentation</a>.</li>
</ul>
</div>
<hr>
<table class="caption-top table">
<colgroup>
<col style="width: 27%">
<col style="width: 37%">
<col style="width: 34%">
</colgroup>
<thead>
<tr class="header">
<th><strong>Method</strong></th>
<th><strong>Advantages</strong></th>
<th><strong>Disadvantages</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Precomputed Metadata</strong></td>
<td>Easy to use; requires minimal setup</td>
<td>Limited to static metadata releases</td>
</tr>
<tr class="even">
<td><strong>REST API</strong></td>
<td>Dynamic queries; programmatic access</td>
<td>Requires scripting</td>
</tr>
<tr class="odd">
<td><strong>Chado PostgreSQL</strong></td>
<td>Comprehensive and direct database access</td>
<td>High setup complexity; requires SQL</td>
</tr>
</tbody>
</table>
<hr>
<ul>
<li><a href="https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help">Fly Atlas 2 database</a>, then report to section <code>Citing  FlyAtlas 2 &amp; Data Availability / Data Availability</code>.</li>
<li><a href="motif.mvls.gla.ac.uk/downloads/FlyAtlas2_2024.01.08.sql">Global SQL databse</a></li>
<li><a href="https://www.ebi.ac.uk/ena/browser/view/PRJEB22205">Original Fly Atlas2 datasets, including both RNASeq and Microarray</a> -&gt; to retrieve all information, likely to use the sql databse rather than excel, where data is aggregated/merged: https://pmc.ncbi.nlm.nih.gov/articles/instance/5753349/bin/gkx976_supp.pdf for an explanation of the SQL schema.</li>
<li><a href="https://github.com/YazBraimah/FlyAtlas2/blob/master/Snakefile">SnakeMake GitHub repository to reproduce all the results + metadata for Fly Atlas 2</a></li>
</ul>



</section>
</div></section><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-zeeberg2004bb" class="csl-entry" role="listitem">
Zeeberg, Barry R., Joseph Riss, David W. Kane, Kimberly J. Bussey, Edward Uchio, W. Marston Linehan, J. Carl Barrett, and John N. Weinstein. 2004. <span>“Mistaken <span>Identifiers</span>: <span>Gene</span> Name Errors Can Be Introduced Inadvertently When Using <span>Excel</span> in Bioinformatics.”</span> <em>BMC Bioinformatics</em> 5 (1): 80. <a href="https://doi.org/10.1186/1471-2105-5-80">https://doi.org/10.1186/1471-2105-5-80</a>.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p><code>FlyAtlas</code> was a compendium of micro-array datasets, and has been outdated by the release of <code>FlyAtlas2</code><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>It appears that the <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#Large_dataset_metadata">Large Dataset metadata file</a> only contains primary and foreign keys for each Drosophila experience and individual, constraining to an external linkout to retrieve such metadata<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>Indeed, most reference-based deconvolution algorithms require an initial stringent feature selection process, requiring both estimation of the mean and individual gene variances.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>Besides, as shown in <span class="citation" data-cites="zeeberg2004bb">(<a href="#ref-zeeberg2004bb" role="doc-biblioref">Zeeberg et al. 2004</a>)</span> demonstrated that up to <span class="math inline">\(30\%\)</span> published papers contain mangled gene names, most of them being induced by Excel spreadsheets’ auto-completion.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>Unfortunately, we can’t directly parse the http(s) repository as proposed in section <a href="#sec-modencode-expression" class="quarto-xref">Section&nbsp;1.3.1</a>. Indeed, most of the recent HTTPs data repositories prevent from navigating their file structures without proper authentication<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p>Of note, both are also available at the transcript level<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7"><p>Returns inconsistent or uncomplete results unfortunately. Had to test the local <a href="https://cran.r-project.org/web/packages/ollamar/index.html"><code>ollamar</code></a> package.**) or the recent <a href="https://github.com/tidyverse/elmer"><code>elmer</code> package</a><a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8"><p><strong>Part of the explanation lies in migration from ArrayExpress to BioStudies API standards, as reported in <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/help#programmatic">Programmatic access ArrayExpress</a></strong><a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn9"><p>All the files that can be downloaded from a single cell experience are detailed in <a href="https://www.ebi.ac.uk/gxa/sc/download">EBI Download Help tabset</a><a href="#fnref9" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn10"><p>Only RPKM and FPKM pre-processing are available on FlyBase.<a href="#fnref10" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn11"><p><a href="https://www.bioconductor.org/packages//2.10/bioc/vignettes/SRAdb/inst/doc/SRAdb.pdf">Vignette for the package</a> and <a href="https://alexslemonade.github.io/2020-june-training/working-with-your-data/retrieve-SRAdb-metadata.nb.html">Recent project use case</a><a href="#fnref11" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn12"><p>In details, the following credentials are expected: <strong>Hostname:</strong>(<code>chado.flybase.org</code>), <strong>Database:</strong>(<code>flybase</code>), <strong>User:</strong>(<code>flybase</code>), no password and <strong>Port:</strong> 5432<a href="#fnref12" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn13"><p>This approach is recommended for high-performance needs.<a href="#fnref13" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>