<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Explore and download sequencing databases: hows-to • DrosophiliaDeconv</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Explore and download sequencing databases: hows-to">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DrosophiliaDeconv</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/DrosophiliaDeconv.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/download.html">Explore and download sequencing databases: hows-to</a></li>
    <li><a class="dropdown-item" href="../articles/preprocess_bulkRNASeq.html">Analyse bulk RNASeq experiences</a></li>
    <li><a class="dropdown-item" href="../articles/preprocess_scRNASeq.html">Analyse single cell RNASeq experiences</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bastienchassagnol/DrosophiliaDeconv/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Explore and download sequencing databases: hows-to</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bastienchassagnol/DrosophiliaDeconv/blob/main/vignettes/download.Rmd" class="external-link"><code>vignettes/download.Rmd</code></a></small>
      <div class="d-none name"><code>download.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="protocol-to-download-and-retrieve-metadata-to-build-reference-datasets">Protocol to download and retrieve metadata to build reference
datasets<a class="anchor" aria-label="anchor" href="#protocol-to-download-and-retrieve-metadata-to-build-reference-datasets"></a>
</h2>
<p><strong>For a nice HTML + quarto presentation, with functional
figure, lst and other float references; citations and nice call-out
boxes, download locally self-embedded HTML file <a href="https://github.com/bastienchassagnol/DrosophiliaDeconv/blob/main/inst/html/download.html" class="external-link">download.html</a>.</strong></p>
<p>======= # Protocol to download and retrieve metadata to build
reference datasets</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Tidyverse libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span> <span class="co"># For regex operations</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span> <span class="co"># vectorial operations</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span> <span class="co"># read and write tsv and csv files</span></span>
<span></span>
<span><span class="co">#  For website exploration</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rvest.tidyverse.org/" class="external-link">rvest</a></span><span class="op">)</span>  <span class="co"># For web scraping and HTML parsing</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.r-universe.dev/jsonlite" class="external-link">jsonlite</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://httr2.r-lib.org" class="external-link">httr2</a></span><span class="op">)</span> <span class="co"># tailored for HTTP/HTTPS requests</span></span>
<span><span class="co"># both packages are better tailored to FTP protocols</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.r-universe.dev/curl" class="external-link">curl</a></span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RCurl</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For SQL parsing and integration with R</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dbi.r-dbi.org" class="external-link">DBI</a></span><span class="op">)</span> <span class="co"># open connection with DBMS</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dbplyr.tidyverse.org/" class="external-link">dbplyr</a></span><span class="op">)</span> <span class="co"># seamless use of dplyr operations with SQL power</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org" class="external-link">RSQLite</a></span><span class="op">)</span> <span class="co"># lightweight DBMS</span></span>
<span></span>
<span></span>
<span><span class="co"># Set global options</span></span>
<span><span class="va">opts_chunk</span><span class="op">$</span><span class="fu">set</span><span class="op">(</span></span>
<span>  echo <span class="op">=</span> <span class="cn">TRUE</span>,         <span class="co"># Show R code in the output</span></span>
<span>  eval <span class="op">=</span> <span class="cn">FALSE</span>,        <span class="co"># By default, do not execute R code</span></span>
<span>  warning <span class="op">=</span> <span class="cn">FALSE</span>,     <span class="co"># Suppress warnings in the output</span></span>
<span>  message <span class="op">=</span> <span class="cn">FALSE</span>,     <span class="co"># Suppress messages in the output</span></span>
<span>  fig.width <span class="op">=</span> <span class="fl">7</span>,       <span class="co"># Default figure width</span></span>
<span>  fig.height <span class="op">=</span> <span class="fl">5</span>,      <span class="co"># Default figure height</span></span>
<span>  fig.align <span class="op">=</span> <span class="st">"center"</span> <span class="co"># Centre-align figures</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set ggplot2 minimal theme for consistency</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="litterature-and-web-scraping">Litterature and web-scraping<a class="anchor" aria-label="anchor" href="#litterature-and-web-scraping"></a>
</h3>
<ul>
<li>RNASeq, tissue-resolved datasets:
<ul>
<li>the <a href="https://wiki.flybase.org/wiki/FlyBase:Drosophila_Online_Resources#Gene_Expression_Databases_and_Tools" class="external-link">FlyBase:Drosophila_Online_Resources</a>
</li>
<li>The Wiki fly page also refers to a a list of tools, Web APIs and
curated pipelines to analyse Drosophila samples.</li>
<li>Most popular and relevant studies are <a href="https://flybase.org/rnaseq/profile_search" class="external-link"><code>modENCODE</code>
tissue and cell lines experiences</a> and <a href="https://academic.oup.com/nar/article/46/D1/D809/4563305" class="external-link"><code>FlyAtlas2</code></a><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;&lt;code&gt;FlyAtlas&lt;/code&gt; was a compendium of micro-array
datasets, and has been outdated by the release of
&lt;code&gt;FlyAtlas2&lt;/code&gt;&lt;/p&gt;"><sup>1</sup></a>.<br>
</li>
</ul>
</li>
<li>scRNASeq datasets:
<ul>
<li><a href="https://wiki.flybase.org/wiki/FlyBase:Drosophila_Online_Resources#Single_Cell_RNA-seq" class="external-link">List
of scRNASeq datasets and tools, at the tissue level</a></li>
<li><a href="https://www.flyrnai.org/tools/single_cell/web/summary" class="external-link">Comprehensive
list of scRNASeq databases, within a tissue level, using the
<code>DRscDB</code> Shiny App</a></li>
</ul>
</li>
<li><a href="https://flybase.org/downloads/bulkdata" class="external-link">Complete list of
Datasets, from FlyAtlas compendium</a></li>
</ul>
<p><strong>All these datasets, with some metadata characterising them,
are reported in Excel table
<code>data/databases_metadata.xlsx</code></strong>, composed of three
sheets:</p>
<ol style="list-style-type: decimal">
<li>
<code>Databases references</code>, enumerating the most relevant
datasets and RNASeq resources.</li>
<li>
<code>DRscDB Databases Single Cell</code>, listing all scRNASeq
datasets having a higher granularity than tissue level (for instance,
characterising the cell subtypes composing the digestive tract).</li>
</ol>
<div class="section level4">
<h4 id="optionnal-retrieve-programmatically-drscdb-single-cell-databases">(Optionnal) Retrieve programmatically DRscDB single cell
databases<a class="anchor" aria-label="anchor" href="#optionnal-retrieve-programmatically-drscdb-single-cell-databases"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Toggle the <a href="https://www.flyrnai.org/tools/single_cell/web/summary" class="external-link"><code>DRscDB</code>
Datasets tab</a>.</li>
<li>Retrieve the metadata dataset using the R web-scraping tool
<code>rvest</code>, and select only the most relevant columns:</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">url_DRscDB</span> <span class="op">&lt;-</span> <span class="st">"https://www.flyrnai.org/tools/single_cell/web/summary"</span></span>
<span><span class="va">irrelevant_columns_DRscDB</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pubmed date"</span>, <span class="st">"dataset ids"</span>, <span class="st">"subset names"</span>,</span>
<span>                               <span class="st">"species"</span>, <span class="st">"species id"</span>,</span>
<span>                               <span class="st">"reads cell"</span>, <span class="st">"genes cell"</span>, <span class="st">"total gene"</span>,</span>
<span>                               <span class="st">"batch correction"</span>, <span class="st">"data in supp table"</span>,</span>
<span>                               <span class="st">"web link available"</span>, <span class="st">"linkout URL"</span><span class="op">)</span></span>
<span><span class="va">metadata_DRscDB</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">read_html</a></span><span class="op">(</span><span class="va">url_DRscDB</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># read a HTML file</span></span>
<span>  <span class="fu"><a href="https://rvest.tidyverse.org/reference/html_table.html" class="external-link">html_table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># from the HTML file, select only Tables (on that website, only one found)</span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">species</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"Drosophila"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Filter on Drosophila species</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">irrelevant_columns_DRscDB</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="retrieve-phenotype-annotation-for-each-dataset">Retrieve phenotype annotation for each dataset<a class="anchor" aria-label="anchor" href="#retrieve-phenotype-annotation-for-each-dataset"></a>
</h3>
<div class="callout-warning" title="Major issues for retrieving Phenotypes">
<ul>
<li>No homogeneous database storage (SQL, Excel, SRA, tsv file, …),
within and in-between studies</li>
<li>While the FlyBase database is comprehensive, does not store properly
annotation data for each tissue, at least for RNASeq samples<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;It appears that the &lt;a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#Large_dataset_metadata" class="external-link"&gt;Large
Dataset metadata file&lt;/a&gt; only contains primary and foreign keys for
each Drosophila experience and individual, constraining to an external
linkout to retrieve such metadata&lt;/p&gt;'><sup>2</sup></a>.</li>
<li>No official and up-to-date BioConductor nor CRAN package to fetch
and explore Drosophila databases</li>
</ul>
<p><strong>Consequence</strong>: specific protocol for retrieving such
phenotypical annotation for each dataset (I enumerate for the top three
popular ones).</p>
</div>
<div class="section level4">
<h4 id="sec-modencode-metadata">The ModENCODE pheno data<a class="anchor" aria-label="anchor" href="#sec-modencode-metadata"></a>
</h4>
<ul>
<li>The easiest way I could find: retrieve metadata from <a href="https://wiki.flybase.org/wiki/FlyBase:ModENCODE_data_at_FlyBase#RNA-Seq_JBrowse_Track_Listing" class="external-link">JBrowse
FlyBase Wiki website</a>
</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ModENCODE_databases</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">read_html</a></span><span class="op">(</span><span class="st">"https://wiki.flybase.org/wiki/FlyBase:ModENCODE_data_at_FlyBase#RNA-Seq_JBrowse_Track_Listing"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rvest.tidyverse.org/reference/html_table.html" class="external-link">html_table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># `trim = FALSE` to ensure missing cells are filled correctly </span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">`Track section`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Expression&gt;RNA-Seq &gt;modENCODE Transcriptomes  &gt;Tissues"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">ModENCODE_databases_formatted</span> <span class="op">&lt;-</span> <span class="va">ModENCODE_databases</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_detect.html" class="external-link">str_detect</a></span><span class="op">(</span><span class="va">`Track section`</span>, <span class="st">"Tissues$"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">`Track section`</span>, <span class="op">-</span> <span class="st">"JBrowse Track Description"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>tissue <span class="op">=</span> <span class="st">"Track name"</span>, FlybaseID<span class="op">=</span><span class="st">"FlyBase Dataset(s)"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>FlybaseID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">FlybaseID</span>,        <span class="co"># split by mE_mRNA, without consuming the separator</span></span>
<span>                   pattern <span class="op">=</span> <span class="st">"(?=mE_mRNA)"</span><span class="op">)</span>, <span class="va">str_subset</span>, <span class="st">"[^ ]"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># remove empty chains </span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/chop.html" class="external-link">unchop</a></span><span class="op">(</span><span class="va">FlybaseID</span><span class="op">)</span> <span class="co"># from compact list of samples per tissue type to tidy format</span></span>
<span></span>
<span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="va">ModENCODE_databases_formatted</span>, </span>
<span>                 <span class="st">"./data/RNASeq/ModENCODE_pheno_data.csv"</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Alternative methods:
<ul>
<li>When pasting a HTML table to an Excel Workbook, use <strong>Paste
with a refreshable web query</strong> option</li>
<li>The <a href="https://flybase.org/rnaseq/profile_search" class="external-link">FlyBase
RNASeq Search profile</a>, clicking on the <code>tissue</code> and
optionally <code>cell line</code> items.</li>
</ul>
</li>
</ul>
<div layout="[[1], [1,1]]">
<div class="float">
<img src="img%2FFlyBase_Profile_Search.png" alt="Screenshot FlyBase Profile Search"><div class="figcaption">Screenshot FlyBase Profile Search</div>
</div>
<div class="float">
<img src="img%2FFlyBase_tissues.png" alt="Screenshot Tissue Samples"><div class="figcaption">Screenshot Tissue Samples</div>
</div>
<div class="float">
<img src="img%2FFlyBase_cell_lines.png" alt="Screenshot Cell Line samples"><div class="figcaption">Screenshot Cell Line samples</div>
</div>
</div>
<ul>
<li>To finalise the annotation, we’re going to use the <a href="https://flybase.github.io/api/swagger-ui/#/HitList/getFlyBaseHitList" class="external-link">Rest
API GET HitList</a>, programmatic access enabled by the FlyBase
website:</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Which individual experiences we want metadata from?</span></span>
<span><span class="va">flybase_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FBlc0000215"</span>, <span class="st">"FBlc0000216"</span>, <span class="st">"FBlc0000217"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a URL with IDs delimited by commas</span></span>
<span><span class="va">base_url</span> <span class="op">&lt;-</span> <span class="st">"https://api.flybase.org/api/v1.0/hitlist/fetch"</span></span>
<span><span class="va">modENCODE_request</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">base_url</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># parameter construction defined here: https://flybase.github.io/api/swagger-ui/#/HitList/getFlyBaseHitList</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="va">flybase_ids</span> <span class="op">|&gt;</span> </span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span>collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the response status, 200 corresponding to a success</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://httr2.r-lib.org/reference/resp_status.html" class="external-link">resp_status</a></span><span class="op">(</span><span class="va">modENCODE_request</span><span class="op">)</span> <span class="op">==</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_has_body</a></span><span class="op">(</span><span class="va">modENCODE_request</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Parse the JSON content into an R object</span></span>
<span>  <span class="va">ModENCODE_metadata</span> <span class="op">&lt;-</span>   <span class="va">modENCODE_request</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_body_string</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/jsonlite/man/fromJSON.html" class="external-link">fromJSON</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">|&gt;</span> </span>
<span>    <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"resultset"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"result"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="co"># convert a HTTP error to a R error</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_status.html" class="external-link">resp_check_status</a></span><span class="op">(</span><span class="va">modENCODE_request</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="the-flyatlas2-pheno-data">The FlyAtlas2 pheno data<a class="anchor" aria-label="anchor" href="#the-flyatlas2-pheno-data"></a>
</h4>
<div class="callout-warning" title="Excel workbook and aggregated dataset">
<p>The <a href="https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help" class="external-link">Excel
Workbook</a> mentioned in the <code>Docs</code> tab section of the
official <code>FlyAtlas2</code> initiative only stores aggregated and
averaged information at the tissue level, rendering it useless for
proper downstream analyses<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Indeed, most reference-based deconvolution algorithms
require an initial stringent feature selection process, requiring both
estimation of the mean and individual gene variances.&lt;/p&gt;"><sup>3</sup></a>.</p>
<p><strong>In conclusion, this method and this dataset is irrelevant for
most downstream analyses</strong><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Besides, as shown in &lt;span class="citation"&gt;(Zeeberg et
al. 2004)&lt;/span&gt; demonstrated that up to
&lt;math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;semantics&gt;&lt;mrow&gt;&lt;mn&gt;30&lt;/mn&gt;&lt;mi&gt;%&lt;/mi&gt;&lt;/mrow&gt;&lt;annotation encoding="application/x-tex"&gt;30\%&lt;/annotation&gt;&lt;/semantics&gt;&lt;/math&gt;
published papers contain mangled gene names, most of them being induced
by Excel spreadsheets’ auto-completion.&lt;/p&gt;'><sup>4</sup></a></p>
</div>
<ul>
<li>
<p>In the FPKM/RPKM file stored in the web encyclopaedia
<code>FlyBase</code>, the column headers identifying uniquely each
sample describe the developmental stage, the sex and the original
tissue. However, it’s not really practical to parse, nor follows an
unique identifier composition. Example:
<code>RNA-Seq_Profile_FlyAtlas2_Adult_Female_Brain_(FBlc0003619)</code>.</p>
<ul>
<li>However, we can use the same REST API trick mentioned in <span class="citation">(<strong>lst-API-FlyBase-modENCODE-metadata?</strong>)</span>
to get some additional information (column <code>title</code>).</li>
</ul>
</li>
<li>
<p>The most challenging method involves re-running the SQL file
provided with the project.</p>
<ol style="list-style-type: decimal">
<li>The name of the SQL file can be retrieved on the <a href="https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help" class="external-link"><code>FlyAtlas2</code>
website</a>, or, in a somehow twisted way with the following R code<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Unfortunately, we can’t directly parse the http(s)
repository as proposed in section &lt;span class="citation"&gt;(&lt;strong&gt;sec-modencode-expression?&lt;/strong&gt;)&lt;/span&gt;.
Indeed, most of the recent HTTPs data repositories prevent from
navigating their file structures without proper authentication&lt;/p&gt;'><sup>5</sup></a>.</li>
</ol>
</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">flyatlas_url</span> <span class="op">&lt;-</span> <span class="st">"https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help"</span></span>
<span><span class="va">webpage_text</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://xml2.r-lib.org/reference/read_xml.html" class="external-link">read_html</a></span><span class="op">(</span><span class="va">flyatlas_url</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rvest.tidyverse.org/reference/html_text.html" class="external-link">html_text</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">flyatlas_files</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract_all</a></span><span class="op">(</span><span class="va">webpage_text</span>, </span>
<span>                                         <span class="st">"motif.mvls.gla.ac.uk/downloads/[^\\s]+"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>From the previous manipulation, two files are available:
motif.mvls.gla.ac.uk/downloads/FlyAtlas2_2024.01.08.sql,
motif.mvls.gla.ac.uk/downloads/FlyAtlas2_gene_data.xlsx, but only the
SQL file is comprehensive, storing gene expressions at the individual
level (and not at the aggregated tissue level). i. The SQL relational
database is described in details in <a href="https://pmc.ncbi.nlm.nih.gov/articles/instance/5753349/bin/gkx976_supp.pdf" class="external-link">Supplementary
FlyAtlas 2 documents, Table S2, Pages 3-4</a>. Of interest, the
<code>Tissue</code>, <code>GeneFPKM</code> and <code>GeneRPM</code>
datasets<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Of note, both are also available at the transcript
level&lt;/p&gt;"><sup>6</sup></a>. ii. Parse the SQL file’s instructions,
keeping SQL lines for building <code>Tissue</code>,
<code>GeneFPKM</code> and <code>GeneRPM</code> datasets. Usually,
building a SQL dataset involves three stages: ensuring the non prior
existence of the dataset (if it’s the case, enforce its deletion),
define the dataset (name and type of features/columns), and finally,
build the dataset itself (indeed, dataset content is hard-coded within
the SQL file):</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sql_instructions</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">read_lines</a></span><span class="op">(</span><span class="st">'data/FlyAtlas2_2024.01.08.sql'</span>,</span>
<span>                                      skip_empty_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># Keep only lines that are NOT empty nor commented (defined by "--")</span></span>
<span>  <span class="co"># stringr::str_subset("^\\s*(--)|/\\*", negate = TRUE) |&gt;</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"^\\s*(--)|/\\*|(UN)?LOCK"</span>, negate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span>collapse <span class="op">=</span> <span class="st">"\r\n"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract_all</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/modifiers.html" class="external-link">regex</a></span><span class="op">(</span><span class="st">"[^;]+;"</span>, multiline <span class="op">=</span> <span class="cn">TRUE</span>,dotall <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                  simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tissue_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CREATE TABLE `Tissue`"</span>, <span class="va">sql_instructions</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">tissue_instructions</span> <span class="op">&lt;-</span> <span class="va">sql_instructions</span><span class="op">[</span><span class="op">(</span><span class="va">tissue_index</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">tissue_index</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">write_lines</a></span><span class="op">(</span><span class="va">tissue_instructions</span>,</span>
<span>                   <span class="st">"data/old_tissue_parsing.sql"</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: lower-roman">
<li>Use modern SQL DBMS. Report to section <span class="citation">(<strong>sec-convert-sql-format?</strong>)</span> for
further details. Major issue with the current Fly Atlas 2 database is
their outdated use of the <code>myISAM</code> DBMS standards, for which
no official converter exists.<br>
</li>
<li>(Optional) Verify your updated SQL instructions can be parsed and
executed using your local DBMS manager like mySQL, or even simpler,
going on <a href="https://www.programiz.com/sql/online-compiler/" class="external-link">SQL
online resources</a>. It turned out that we had to remove an uniqueness
constraint on <code>Tissue.Abbreviation</code> feature to run the SQL
code.</li>
<li>Execute the modern SQL file, store transiently the output to a
SQLite database, collect and export the results as a standard csv
file.</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create SQLite database connection</span></span>
<span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dbi.r-dbi.org/reference/dbConnect.html" class="external-link">dbConnect</a></span><span class="op">(</span><span class="fu">RSQLite</span><span class="fu">::</span><span class="fu"><a href="https://rsqlite.r-dbi.org/reference/SQLite.html" class="external-link">SQLite</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"data/FlyAtlas2_tissue.sqlite"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sql_statements</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">read_lines</a></span><span class="op">(</span><span class="st">"data/new_tissue.sql"</span>, </span>
<span>                                    skip_empty_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"^\\s*(--)|/\\*|(UN)?LOCK"</span>, negate <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span>collapse <span class="op">=</span> <span class="st">"\r\n"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract_all</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/modifiers.html" class="external-link">regex</a></span><span class="op">(</span><span class="st">"[^;]+;"</span>, multiline <span class="op">=</span> <span class="cn">TRUE</span>,dotall <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>                  simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Tip: since dbExecute can only perform one query after the other, we use a for-loop to perform sequential SQL queries</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">statement</span> <span class="kw">in</span> <span class="va">sql_statements</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://dbi.r-dbi.org/reference/dbExecute.html" class="external-link">dbExecute</a></span><span class="op">(</span><span class="va">con</span>, <span class="va">statement</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># testthat::expect_contains(dbListTables(con), "Tissue")</span></span>
<span><span class="co"># DBI::dbRemoveTable(con, "Tissue")</span></span>
<span></span>
<span><span class="co"># Export the tissue dataset to a standard CSV format</span></span>
<span><span class="va">tissue_fly_atlas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html" class="external-link">tbl</a></span><span class="op">(</span><span class="va">con</span>, <span class="st">"Tissue"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="va">tissue_fly_atlas</span>,</span>
<span>                 <span class="st">"data/pheno_data_flyatlas2.csv"</span><span class="op">)</span></span>
<span><span class="co"># Close the connection once done</span></span>
<span><span class="fu"><a href="https://dbi.r-dbi.org/reference/dbDisconnect.html" class="external-link">dbDisconnect</a></span><span class="op">(</span><span class="va">con</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="sec-convert-sql-format">(Optionnal) Convert old SQL version 5 to modern SQL version 9<a class="anchor" aria-label="anchor" href="#sec-convert-sql-format"></a>
</h4>
<p>While functionalities going along the <code>DBI</code> R package
would theoretically enable us to execute SQL instructions directly from
R, it turns out that the SQL instructions are tailored to <a href="https://en.wikipedia.org/wiki/MyISAM" class="external-link">MyISAM DBMS software</a>,
outdated from 2009, and can not be processed directly by the
<code>DBI</code> wrapper. Besides, <code>DBI</code> works best with
<code>SQLite</code> DBMS, and the FlyBase SQL file has not been written
to that end. To conclude, we have to convert the SQL FlyBase file to
modern SQL standards (current version: <code>SQL Server 2022</code>)
while ensuring compatibility with <code>SQLite</code>.</p>
<p>No dedicated tools, up to my knowledge, have been developed in R to
that end. So definitely, the best method, while being intrinsically
stochastic, would consist of pairing a LLM using API REST integration
with <em>prompt engineering</em>. I propose two code snippets below with
a personalised engineered prompt, using <a href="https://github.com/mlverse/chattr" class="external-link"><code>chattr</code></a>, see
<span class="citation">(<strong>lst-chattr-sql?</strong>)</span> and <a href="https://www.r-bloggers.com/2024/06/large-language-models-llms-in-r-with-tidychatmodels/" class="external-link"><code>tidychatmodels</code></a>,
see <span class="citation">(<strong>lst-tidychatmodels-sql?</strong>)</span>,
respectively<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;Returns inconsistent or uncomplete results
unfortunately. Had to test the local &lt;a href="https://cran.r-project.org/web/packages/ollamar/index.html" class="external-link"&gt;&lt;code&gt;ollamar&lt;/code&gt;&lt;/a&gt;
package.**) or the recent &lt;a href="https://github.com/tidyverse/elmer" class="external-link"&gt;&lt;code&gt;elmer&lt;/code&gt;
package&lt;/a&gt;&lt;/p&gt;'><sup>7</sup></a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># remotes::install_github("mlverse/chattr")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mlverse/chattr" class="external-link">chattr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># parameter configuration and saving</span></span>
<span><span class="fu">chattr_use</span><span class="op">(</span><span class="st">"copilot"</span><span class="op">)</span></span>
<span><span class="fu">chattr_defaults</span><span class="op">(</span>max_data_files <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                include_doc_contents <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                max_data_frames <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                yaml_file <span class="op">=</span> <span class="st">"vignettes/chattr.yml"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">new_sql</span> <span class="op">&lt;-</span> <span class="fu">chattr</span><span class="op">(</span>prompt <span class="op">=</span> <span class="st">"Please convert the SQL file &lt;old-sql-location&gt;, written with SQL version 5, to SQL Server 2022 syntax, while making it compliant with the SQLite DBMS constraints."</span>,</span>
<span>       prompt_build <span class="op">=</span> <span class="st">"Specifically, focus on the following: Change data types to match SQLite standards (e.g., TINYINT to INTEGER, ENUM to TEXT). Replace AUTO_INCREMENT with AUTOINCREMENT. Remove any unsupported SQL syntax in SQLite. Write the resulting modernised SQL file to `data/new_tissue_global.sql`"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># devtools::install_github("AlbertRapp/tidychatmodels")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tidychatmodels</span><span class="op">)</span></span>
<span></span>
<span><span class="va">chat_openai</span> <span class="op">&lt;-</span> <span class="fu">create_chat</span><span class="op">(</span><span class="st">'openai'</span>, <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">'OAI_DEV_KEY'</span><span class="op">)</span><span class="op">)</span><span class="op">|&gt;</span></span>
<span>  <span class="fu">add_model</span><span class="op">(</span><span class="st">'gpt-3.5-turbo'</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># add_params(temperature = 0.5, max_tokens = 100) |&gt;</span></span>
<span>  <span class="co"># define the general AI agent system</span></span>
<span>  <span class="fu">add_message</span><span class="op">(</span></span>
<span>    role <span class="op">=</span> <span class="st">'system'</span>,</span>
<span>    message <span class="op">=</span> <span class="st">'You are a chatbot that only returns SQL code, whose syntax is</span></span>
<span><span class="st">    compatible with SQL Server 2022 and SQLite DBMS.'</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># create your customised prompt</span></span>
<span>  <span class="fu">add_message</span><span class="op">(</span></span>
<span>    role <span class="op">=</span> <span class="st">'user'</span>,</span>
<span>    message <span class="op">=</span> <span class="st">'Convert old sql file &lt;old-sql-path&gt; to recent SQL code,</span></span>
<span><span class="st">    write the output to &lt;new-sql-path&gt;. Specifically: Update data types to align</span></span>
<span><span class="st">    with SQLite standards, such as replacing TINYINT with INTEGER,</span></span>
<span><span class="st">    ENUM with TEXT, and other incompatible types. Change AUTO_INCREMENT to</span></span>
<span><span class="st">    AUTOINCREMENT for primary keys. Remove syntax elements unsupported in SQLite,</span></span>
<span><span class="st">    like full-text indexes, foreign key constraints outside PRAGMA, or stored procedures.</span></span>
<span><span class="st">    Where conflicts between SQL Server and SQLite arise, prioritize SQLite compliance.</span></span>
<span><span class="st">    Output the transformed SQL file as a set of discrete statements, one per line,</span></span>
<span><span class="st">    removing empty or commented lines for clarity. Do not add any empty lines nor comments. Do not add uniqueness constraints. Convert the entire SQL file, without using further rows argument.'</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">chat_user_result</span> <span class="op">&lt;-</span> <span class="va">chat_openai</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">perform_chat</span><span class="op">(</span><span class="op">)</span><span class="op">|&gt;</span></span>
<span>  <span class="fu">extract_chat</span><span class="op">(</span>silent <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>And finally, an alternative SQL method, which mixes a bunch of regex
expressions and <code>tidyverse</code> manipulations, while guaranteeing
a determined output:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tissue_colnames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CREATE TABLE `Tissue`"</span>, <span class="va">sql_instructions</span>,</span>
<span>                        fixed <span class="op">=</span> <span class="cn">TRUE</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract_all</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"(?&lt;=\\`)[[:alnum:]]+(?=\\`)"</span>,</span>
<span>                           simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="st">"Tissue"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tissue_inputs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"INSERT INTO `Tissue`"</span>, <span class="va">sql_instructions</span>,</span>
<span>                      fixed <span class="op">=</span> <span class="cn">TRUE</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract_all</a></span><span class="op">(</span><span class="st">"(?&lt;=\\()([^)]+)(?=\\),*)"</span>,</span>
<span>                                                 simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="st">"'"</span>, <span class="st">""</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span>\<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">","</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>               <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/list_c.html" class="external-link">list_rbind</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span>nm <span class="op">=</span> <span class="va">tissue_colnames</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="va">tissue_inputs</span>,</span>
<span>                 file <span class="op">=</span> <span class="st">"data/pheno_data_flyatlas2.csv"</span>,</span>
<span>                 escape <span class="op">=</span> <span class="st">"none"</span>, quote <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="sec-fca-pheno-data">The single cell Fly Cell Atlas pheno data<a class="anchor" aria-label="anchor" href="#sec-fca-pheno-data"></a>
</h4>
<p>One possibility to retrieve phenotype data would consist of using the
<a href="https://www.bioconductor.org/packages/release/bioc/manuals/ArrayExpress/man/ArrayExpress.pdf" class="external-link"><code>BioConductor ArrayExpress</code></a>
package. Unfortunately, <strong>it does not seem updated anymore, and
was indeed unable to download and parse the two FCA experiences
(actually, even the examples from the package documentation can’t be
executed anymore)</strong><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;strong&gt;Part of the explanation lies in migration from
ArrayExpress to BioStudies API standards, as reported in &lt;a href="https://www.ebi.ac.uk/biostudies/arrayexpress/help#programmatic" class="external-link"&gt;Programmatic
access ArrayExpress&lt;/a&gt;&lt;/strong&gt;&lt;/p&gt;'><sup>8</sup></a>. Typical pipeline is illustrated below:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#if (!requireNamespace("BiocManager", quietly = TRUE))</span></span>
<span><span class="co">#    install.packages("BiocManager")</span></span>
<span><span class="co"># BiocManager::install("ArrayExpress")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ArrayExpress</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download both processed and raw datasets</span></span>
<span><span class="va">ae_files</span> <span class="op">&lt;-</span> <span class="fu">getAE</span><span class="op">(</span><span class="st">"E-MTAB-10628"</span>, type <span class="op">=</span> <span class="st">"full"</span>, path <span class="op">=</span> <span class="st">"data/Fly_Cell_Atlas_10x/"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Build a an ExpressionSet from the raw data</span></span>
<span><span class="va">ae_data_raw</span> <span class="op">&lt;-</span> <span class="fu">ae2bioc</span><span class="op">(</span>mageFiles <span class="op">=</span> <span class="va">ae_files</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Build a an ExpressionSet from the processed data</span></span>
<span><span class="va">ae_data</span> <span class="op">&lt;-</span> <span class="fu">procset</span><span class="op">(</span><span class="va">ae_files</span>, <span class="fu">getcolproc</span><span class="op">(</span><span class="va">ae_files</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># equivalent to run ArrayExpress function</span></span>
<span><span class="va">ae_data</span> <span class="op">&lt;-</span> <span class="fu">ArrayExpress</span><span class="op">(</span><span class="st">"E-MTAB-10628"</span>, path <span class="op">=</span> <span class="st">"data/Fly_Cell_Atlas_10x/"</span>,</span>
<span>                        save <span class="op">=</span> <span class="cn">TRUE</span>, dataCols <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># use methods from the Biobase::ExpressionSet object to retrieve respectively </span></span>
<span><span class="va">pheno_data</span> <span class="op">&lt;-</span> <span class="fu">phenoData</span><span class="op">(</span><span class="va">ae_data</span><span class="op">)</span> <span class="co"># phenotype data</span></span>
<span><span class="va">experiment_data</span> <span class="op">&lt;-</span> <span class="fu">experimentData</span><span class="op">(</span><span class="va">ae_data</span><span class="op">)</span> <span class="co"># MIAME experiment annotation (metadata)</span></span>
<span><span class="va">gene_annotation</span> <span class="op">&lt;-</span> <span class="fu">fData</span><span class="op">(</span><span class="va">ae_data</span><span class="op">)</span> <span class="co"># gene feature annotation</span></span>
<span><span class="va">expression_matrix</span> <span class="op">&lt;-</span> <span class="fu">exprs</span><span class="op">(</span><span class="va">ae_data</span><span class="op">)</span> <span class="co"># expression matrix data</span></span>
<span></span>
<span><span class="co"># worth noted that the ExpressionSet object is well tailored for bulk micro-array </span></span>
<span><span class="co"># and RNASeq, but not really for singlecell RNASeq experiences</span></span></code></pre></div>
<p>Alternatively, you can download both <code>10X</code> and
<code>Smartseq2</code> experiences and expression data at the count
level (raw and after applying <em>TPM normalisation</em>) using the
following <a href="https://www.ebi.ac.uk/gxa/sc/experiments?species=%22drosophila+melanogaster%22&amp;experimentProjects=%22Fly+Cell+Atlas%22&amp;kingdom=%22Animals%22" class="external-link">Single
cell Expression Atlas web query</a>, then crossing out both datasets,
and finally clicking on the <code>Download 2 entries</code> button
feature.</p>
<p>You can perform it simultaneously for both datasets, or one dataset
after the other, as illustrated in <span class="citation">(<strong>lst-array-express-https-protocol?</strong>)</span>
for the <code>https</code> protocol (only a subsetting of files is
available through this protocol) and the <code>ftp</code> protocol,
<span class="citation">(<strong>lst-array-express-ftp-protocol?</strong>)</span><a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;All the files that can be downloaded from a single cell
experience are detailed in &lt;a href="https://www.ebi.ac.uk/gxa/sc/download" class="external-link"&gt;EBI Download Help
tabset&lt;/a&gt;&lt;/p&gt;'><sup>9</sup></a>:
<!--# write a function to automate all these download queries and checks + use req_perform_parallel to proceed all downloads in parallel for enhanced speed combined with a progress bar --></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download/zip?fileType=experiment-metadata&amp;accessKey="</span></span>
<span><span class="co"># "https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download?fileType=experiment-design&amp;accessKey="</span></span>
<span><span class="co">#   https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download/zip?fileType=normalised&amp;accessKey=</span></span>
<span><span class="co">#   https://www.ebi.ac.uk/gxa/sc/experiment/E-MTAB-10628/download/zip?fileType=quantification-raw&amp;accessKey=</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">base_url</span> <span class="op">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/gxa/sc/experiment/"</span></span>
<span><span class="va">array_identifer</span> <span class="op">&lt;-</span> <span class="st">"E-MTAB-10628"</span> <span class="co"># the SmartSeq identifier</span></span>
<span><span class="co"># array_identifer &lt;- "E-MTAB-10519" # the 10X identifier</span></span>
<span><span class="co"># we took the example of the metadata zip file, containing idf and sdrf files</span></span>
<span><span class="co"># should be extended to normalised and raw expression files + experiment design</span></span>
<span><span class="va">metadata_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"./data/Fly_Cell_Atlas_10x"</span>,</span>
<span>                           <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">array_identifer</span>, <span class="st">"-experiment-metadata"</span>, <span class="st">".zip"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fca_request</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">base_url</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="va">array_identifer</span>, <span class="st">"download"</span>, <span class="st">"zip"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_query</a></span><span class="op">(</span>fileType<span class="op">=</span><span class="st">"experiment-metadata"</span>, accessKey<span class="op">=</span><span class="st">""</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_progress.html" class="external-link">req_progress</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"down"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">metadata_path</span><span class="op">)</span> </span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_has_body</a></span><span class="op">(</span><span class="va">fca_request</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://httr2.r-lib.org/reference/resp_status.html" class="external-link">resp_is_error</a></span><span class="op">(</span><span class="va">fca_request</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">zip</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/zip/reference/unzip.html" class="external-link">unzip</a></span><span class="op">(</span><span class="va">metadata_path</span>, </span>
<span>             overwrite <span class="op">=</span> <span class="cn">TRUE</span>, exdir <span class="op">=</span> <span class="st">"./data/Fly_Cell_Atlas_10x"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="va">metadata_path</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">base_url</span> <span class="op">&lt;-</span> <span class="st">"ftp.ebi.ac.uk/pub/databases/microarray/data/atlas/sc_experiments"</span></span>
<span><span class="va">array_identifier</span> <span class="op">&lt;-</span> <span class="st">"E-MTAB-10628"</span></span>
<span><span class="va">ftp_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">base_url</span>, <span class="st">"/"</span>, <span class="va">array_identifier</span>, <span class="st">"/"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List all files in the FTP directory</span></span>
<span><span class="va">ftp_filenames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/curl/man/curl_fetch.html" class="external-link">curl_fetch_memory</a></span><span class="op">(</span><span class="va">ftp_url</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"content"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rawConversion.html" class="external-link">rawToChar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="co"># Split by rows (newline character)</span></span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span> pattern <span class="op">=</span> <span class="st">"\r\n"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"[^ ]"</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># remove empty lines |&gt;</span></span>
<span>  <span class="co"># matches one or more characters that are not whitespace, to the end of the string.</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span> pattern <span class="op">=</span> <span class="st">"[^\\s]+$"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">idf_index</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"idf"</span>, <span class="va">ftp_URLs</span><span class="op">)</span></span>
<span><span class="va">ftp_URLs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">ftp_url</span>, <span class="va">ftp_filenames</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">ftp_URLs</span><span class="op">[</span><span class="va">idf_index</span><span class="op">]</span>,</span>
<span>              destfile <span class="op">=</span> <span class="va">ftp_filenames</span><span class="op">[</span><span class="va">idf_index</span><span class="op">]</span>,</span>
<span>              method <span class="op">=</span> <span class="st">"auto"</span><span class="op">)</span></span></code></pre></div>
<div class="callout-tip" collapse="true" title="ArrayExpress datasets">
<p>ArrayExpress datasets are traditionally organized into the following
key file types, ordered by increasing level of detail:</p>
<ul>
<li>
<p><strong>Investigation Description Format (IDF)</strong> provides
a high-level overview of the experiment.</p>
<ul>
<li>
<strong>Title and Description</strong>: Summary of the experiment’s
purpose.</li>
<li>
<strong>Design</strong> and <strong>Protocols</strong>: Experimental
design and factors studied (e.g., treatments, time points).</li>
<li>
<strong>Contacts</strong>: Information about the authors or
submitters.</li>
</ul>
</li>
<li><p><strong>Sample and Data Relationship Format (SDRF)</strong>: the
phenotype data, detailing the mapping between <strong>Sample
Identifiers</strong>, <strong>Factor Values</strong> and <strong>Data
File Names</strong></p></li>
<li><p><strong>Array Design Format (ADF)</strong> used to provide the
design of microarray platforms used in the experiment. Equivalent to
<strong>featureData</strong>, setting up the mapping between the genes
sampled and their modern and standard equivalents.</p></li>
<li>
<p>Expression matrices, as:</p>
<ul>
<li>
<strong>Raw Data Files</strong> (from FASTQ (for sequencing data) to
CEL (for Affymetrix arrays))</li>
<li>
<strong>Processed Data Files</strong> (usually counts, after
normalization or transformation)</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<div class="section level3">
<h3 id="retrieve-transcriptomic-annotation">Retrieve transcriptomic annotation<a class="anchor" aria-label="anchor" href="#retrieve-transcriptomic-annotation"></a>
</h3>
<div class="section level4">
<h4 id="sec-modencode-expression">The ModENCODE expression matrix<a class="anchor" aria-label="anchor" href="#sec-modencode-expression"></a>
</h4>
<p>The current RPKM database storing all bulk RNASeq experiences on the
comprehensive FlyBase data collection can be retrieved <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#RNA-Seq_RPKM_values_.28gene_rpkm_report_fb_.2A.tsv.gz.29" class="external-link">here</a>.</p>
<p>Alternatively, we can explore the FTP repository, sub-setting the
files’ selection to those mentioning <code>gene_rpkm</code> explicitly
in their name<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Only RPKM and FPKM pre-processing are available on
FlyBase.&lt;/p&gt;"><sup>10</sup></a>
<!-- Develop a custom R function to parse files retrieved from a FTP repository --></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># FTP URL for the repository</span></span>
<span><span class="va">ftp_url</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.flybase.net/releases/current/precomputed_files/genes/"</span></span>
<span><span class="co"># List all files in the FTP directory</span></span>
<span><span class="va">ftp_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/curl/man/curl_fetch.html" class="external-link">curl_fetch_memory</a></span><span class="op">(</span><span class="va">ftp_url</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"content"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rawConversion.html" class="external-link">rawToChar</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="co"># Split by rows (newline character)</span></span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span> pattern <span class="op">=</span> <span class="st">"\r\n"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"[^ ]"</span><span class="op">)</span> <span class="co"># remove empty lines</span></span>
<span></span>
<span><span class="co"># Split each row by columns (tab character) and convert to a data.frame</span></span>
<span><span class="va">ftp_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, </span>
<span>                               <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">ftp_files</span>, pattern <span class="op">=</span> <span class="st">"[[:blank:]]{2,}"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ftp_files</span> <span class="op">&lt;-</span> <span class="va">ftp_files</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate_wider_delim.html" class="external-link">separate_wider_delim</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="st">"V4"</span>,</span>
<span>                              delim <span class="op">=</span> <span class="st">" "</span>, </span>
<span>                              too_many <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"drop"</span><span class="op">)</span>, <span class="co"># silently remove file renaming</span></span>
<span>                              names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"size"</span>, <span class="st">"month"</span>, <span class="st">"day"</span>, <span class="st">"hour"</span>, <span class="st">"filename"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rnaseq_ftp_files</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="va">ftp_files</span><span class="op">$</span><span class="va">filename</span>, </span>
<span>                                         pattern <span class="op">=</span> <span class="st">"gene_rpkm"</span><span class="op">)</span></span></code></pre></div>
<p>In that case, only 0 store gene expression data, namely :</p>
<ul>
<li>NA: <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#RNA-Seq_RPKM_values_matrix_.28gene_rpkm_matrix_fb_.2A.tsv.gz.29" class="external-link">Table
description</a>. <strong>Note that <code>ModENCODE</code> data is
provided with RPKM pre-processed RNA-Seq expression values, while
<code>FlyAtlas2</code> data has been normalised to FPKM
units.</strong>
</li>
<li>NA: <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview#RNA-Seq_RPKM_values_.28gene_rpkm_report_fb_.2A.tsv.gz.29" class="external-link">Table
description</a>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="the-flyatlas2-expression-matrix">The FlyAtlas2 expression matrix<a class="anchor" aria-label="anchor" href="#the-flyatlas2-expression-matrix"></a>
</h4>
</div>
<div class="section level4">
<h4 id="the-fly-cell-atlas-expression-matrix">The Fly Cell Atlas expression matrix<a class="anchor" aria-label="anchor" href="#the-fly-cell-atlas-expression-matrix"></a>
</h4>
<p>Depending on the level of resolution, the new <code>Biostudies</code>
website, aiming at replacing the <code>ArrayExpress</code> web server,
provides two APIs:</p>
<ol style="list-style-type: decimal">
<li><p>To download files at the expression level (counts), both
normalised and raw counts, we already detailed a programmatic process in
<span class="citation">(<strong>sec-fca-pheno-data?</strong>)</span>.
Note that in addition to the protocol detailed in previous section, you
may alternatively use <code>FTP</code>, <code>curl</code>,
<code>Aspera</code> or <code>Globus</code> software. However, these
tools usually require to know in advance all file locations and paths to
be downloaded. All these tools are further detailed in <a href="https://www.ebi.ac.uk/biostudies/help#" class="external-link">Biostudies
documention</a>.</p></li>
<li><p>The ENA Portal API (see <span class="citation">(<strong>lst-ena-ebi-api?</strong>)</span> for
programmatic access) can be used to retrieve raw files, at the sequence
level (in other words, the bio-informatician job, with possibility to
retrieve both FASTQ and BAM files, the latter corresponding to sequences
already aligned against a reference genome).</p></li>
</ol>
<ul>
<li>It requires to retrieve the corresponding ENA or Project accession
number (usually, the <code>PRJ...</code> is the main identifier, while
the <code>ERP...</code> is the second, optional identifier).
<code>E-MTAB-10628</code> is paired with <code>PRJEB45993</code> project
(ENA identifier: <code>ERP130174</code>) and, while
<code>E-MTAB-10519</code> is paired with <code>PRJEB45570</code> (ENA
identifier: <code>ERP129698</code>).</li>
<li>More documentation details can be found <a href="https://ena-docs.readthedocs.io/en/latest/retrieval/programmatic-access.html" class="external-link">here</a>,
and an interactive swagger UI is reported <a href="https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html" class="external-link">here</a>.</li>
<li>Without this mapping information, you can browse it on the
<code>Biostudies</code> website (report to <span class="citation">(<strong>fig-ERP-MAGE?</strong>)</span> for a
snapshot), or programmatically by either parsing the <code>idf</code>
file or retrieving the json file associated with the Biostudies online
accession (see <span class="citation">(<strong>lst-mtab-ena-mapping?</strong>)</span>).</li>
</ul>
<div class="float" id="fig-ERP-MAGE">
<img src="img%2FE-MTAB-10519-ERR-correspondance.png" alt="ERP and ExpressionSet MAGE mapping"><div class="figcaption">ERP and ExpressionSet MAGE mapping</div>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># interactive swagger ui https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html#/Search%20%26%20Discovery/getResults</span></span>
<span><span class="va">url_base</span> <span class="op">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/ena/portal/api"</span></span>
<span><span class="co"># all available result types for a given study:</span></span>
<span><span class="va">ena_avalaible_datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">url_base</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="st">"results"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_query</a></span><span class="op">(</span>format<span class="op">=</span><span class="st">"tsv"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_body_string</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span>show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># of interest for raw files, the `read_run` is certainly the most complete</span></span>
<span><span class="co"># request for retrieving all potential fields for the read_run dataset</span></span>
<span><span class="co"># interactive: https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html#/Search%20%26%20Discovery/getReturnFields</span></span>
<span><span class="va">read_run_fields</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">url_base</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="st">"returnFields"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_query</a></span><span class="op">(</span>dataPortal<span class="op">=</span><span class="st">"ena"</span>, result<span class="op">=</span><span class="st">"read_run"</span>, format<span class="op">=</span><span class="st">"tsv"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_body_string</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span>show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># and finally, for a given project, the direct FTP links to fastq or BAM files</span></span>
<span><span class="co"># interactive: https://www.ebi.ac.uk/ena/portal/api/swagger-ui/index.html#/Search%20%26%20Discovery/fileReport</span></span>
<span><span class="va">PRJEB45993_BAM_FASTQ_URLs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">url_base</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="st">"filereport"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_query</a></span><span class="op">(</span>accession<span class="op">=</span><span class="st">"PRJEB45993"</span>,</span>
<span>                result<span class="op">=</span><span class="st">"read_run"</span>,</span>
<span>                fields<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"study_accession"</span>, <span class="st">"experiment_accession"</span>,<span class="st">"run_accession"</span>,</span>
<span>                         <span class="st">"fastq_ftp"</span>, <span class="st">"fastq_md5"</span>,</span>
<span>                         <span class="st">"bam_ftp"</span>,<span class="st">" bam_md5"</span><span class="op">)</span>,</span>
<span>                format<span class="op">=</span><span class="st">"tsv"</span>,</span>
<span>                download <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                .multi <span class="op">=</span> <span class="st">"comma"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="co"># for whatever reason, piping all these operations lead to time out. </span></span>
<span><span class="va">read_run_fields</span> <span class="op">&lt;-</span> <span class="va">read_run_fields</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_body_string</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span>show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># number of rows:</span></span>
<span><span class="co"># https://www.ebi.ac.uk/ena/portal/api/filereportcount?result=read_run&amp;accession=PRJEB45993&amp;format=tsv</span></span>
<span></span>
<span><span class="va">read_run_rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">url_base</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="st">"filereportcount"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_query</a></span><span class="op">(</span>accession<span class="op">=</span><span class="st">"PRJEB45993"</span>,</span>
<span>                result<span class="op">=</span><span class="st">"read_run"</span>,</span>
<span>                format<span class="op">=</span><span class="st">"json"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_body_json</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"count"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">testthat</span><span class="fu">::</span><span class="fu"><a href="https://testthat.r-lib.org/reference/equality-expectations.html" class="external-link">expect_equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">read_run_fields</span><span class="op">)</span>, </span>
<span>                       <span class="va">read_run_rows</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">url_base</span> <span class="op">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/biostudies"</span></span>
<span><span class="va">array_identifier</span> <span class="op">&lt;-</span> <span class="st">"E-MTAB-10519"</span></span>
<span><span class="va">ERP_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">url_base</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="st">"files"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="va">array_identifier</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">array_identifier</span>, <span class="st">".json"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_body_json</a></span><span class="op">(</span>simplifyVector <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"section"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="st">"links"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/pluck.html" class="external-link">pluck</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="st">"url"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"^ERP"</span><span class="op">)</span></span></code></pre></div>
<p>Of note, without REST API programmatic access, the <a href="https://www.bioconductor.org/packages//2.10/bioc/vignettes/SRAdb/inst/doc/SRAdb.pdf" class="external-link"><strong>BioConductor
SRAdb package</strong></a> provides programmatic access to metadata from
the Sequence Read Archive (SRA) and ENA (European Nucleotide Archive)
from R<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;&lt;a href="https://www.bioconductor.org/packages//2.10/bioc/vignettes/SRAdb/inst/doc/SRAdb.pdf" class="external-link"&gt;Vignette
for the package&lt;/a&gt; and &lt;a href="https://alexslemonade.github.io/2020-june-training/working-with-your-data/retrieve-SRAdb-metadata.nb.html" class="external-link"&gt;Recent
project use case&lt;/a&gt;&lt;/p&gt;'><sup>11</sup></a>.</p>
<p>However, it relies on a strong SQLite dependency, has not been
maintained for more than two years, and requires prior installation of
the SQL metadata database, implying strong storage and memory-resource.
Besides, search of the database has been streamlined by the development
of the <a href="https://www.ebi.ac.uk/ena/browser/api/swagger-ui/index.html" class="external-link">ENA
Browser API</a>.</p>
<p>Other methods for non-programmers are reported in <a href="https://ena-docs.readthedocs.io/en/latest/retrieval/file-download.html#downloading-files" class="external-link">Download
Files ENA Documentation</a>, with popular tools like
<code>ENA Browser</code>, <code>ENA FTP Downloader GUI tool</code>,
<code>Globus</code> or <code>Aspera</code>. While practical for
downloading thousands of files, with dedicated facilities to manage
parallel downloading or error handling, they require custom and often
admin rights’ installation, depending on third-party software. Besides,
they’re not really practical for enforcing FAIR and reproducibility
guidelines.</p>
<div class="callout-tip" title="MD5 role" collapse="true">
<p>You may have noticed in <span class="citation">(<strong>lst-ena-ebi-api?</strong>)</span> that we have
collected both original raw files and their MD5 peers. Indeed, the
<strong>MD5 file</strong> stores a 32-character hexadecimal string
generated that guarantees data integrity (check for file’s alterations
or corruptions, partial downloading, …). Any mismatch in checksums
indicates that the file has been corrupted. The easiest way to verify
file integrity is certainly by looping over each pair of original file -
MD5 file, using the <code>digest</code> package:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eddelbuettel/digest" class="external-link">digest</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define the file path and expected MD5 checksum</span></span>
<span><span class="va">file_path</span> <span class="op">&lt;-</span> <span class="st">"path/to/your/file.txt"</span>  <span class="co"># Replace with your file's path</span></span>
<span><span class="va">expected_md5</span> <span class="op">&lt;-</span> <span class="st">"d41d8cd98f00b204e9800998ecf8427e"</span>  <span class="co"># Replace with the provided checksum</span></span>
<span></span>
<span><span class="co"># Compute the MD5 checksum of the file</span></span>
<span><span class="va">calculated_md5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/digest/man/digest.html" class="external-link">digest</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">file_path</span>, algo <span class="op">=</span> <span class="st">"md5"</span>, serialize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># testthat::expect_equal(calculated_md5, expected_md5)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="optionnal-ftp-access-on-biostudies-and-not-ena">(Optionnal) FTP access on Biostudies (and not ENA)<a class="anchor" aria-label="anchor" href="#optionnal-ftp-access-on-biostudies-and-not-ena"></a>
</h5>
<p>We detailed in the following code snippet:</p>
<ol style="list-style-type: decimal">
<li><p>the <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/help#programmatic" class="external-link">Biostudies
API</a> to retrieve the most comprehensive metadata file for a given
study, the latter listing also differences with the previous
ArrayExpress web server.</p></li>
<li><p>A <a href="https://www.rdocumentation.org/packages/RCurl/versions/1.98-1.16/topics/getURL" class="external-link"><code>RCurl</code></a>
simplified code to list and download all files along with their
properties (such as size or access rights). Yet, the Biostudies
repository is usually much less comprehensive and organised as the
ENA/SRA web server, while exhibiting the following odd path composition:
<code>ftp.sra.ebi.ac.uk/vol1/&lt;file-type&gt;:err|fastq|run/&lt;accession-prefix&gt;/&lt;00-last-digit-of-full-accession&gt;/&lt;full-accession&gt;/</code>,
described thoroughly in <a href="https://ena-docs.readthedocs.io/en/latest/retrieval/file-download/sra-ftp-structure.html" class="external-link">SRA
FTP Structure documentation</a>, and preventing downloading the whole
study without access to metadata and file localisations.</p></li>
</ol>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1) not that relevant, too many fields to process, without clear structure</span></span>
<span><span class="va">url_base</span> <span class="op">&lt;-</span> <span class="st">"https://www.ebi.ac.uk/biostudies/api/v1/studies"</span></span>
<span><span class="va">array_identifier</span> <span class="op">&lt;-</span> <span class="st">"E-MTAB-10519"</span></span>
<span><span class="va">ena_detailled_information</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://httr2.r-lib.org/reference/request.html" class="external-link">request</a></span><span class="op">(</span><span class="va">url_base</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_url.html" class="external-link">req_url_path_append</a></span><span class="op">(</span><span class="va">array_identifier</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/req_perform.html" class="external-link">req_perform</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://httr2.r-lib.org/reference/resp_body_raw.html" class="external-link">resp_body_json</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2) easiest process for listing all files in a FTP repo using RCurl</span></span>
<span><span class="va">base_url</span> <span class="op">&lt;-</span> <span class="st">'ftp.ebi.ac.uk/biostudies'</span></span>
<span><span class="va">study_type</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="va">array_identifier</span>, <span class="st">"^[[[:alpha:]]-]+"</span><span class="op">)</span></span>
<span><span class="va">study_suffix</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_extract.html" class="external-link">str_extract</a></span><span class="op">(</span><span class="va">array_identifier</span>, <span class="st">"[[:digit:]]{3}$"</span><span class="op">)</span></span>
<span><span class="co"># remove the Files/ if you just wish to access general metadata</span></span>
<span><span class="va">biostudies_ftp_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">base_url</span>, <span class="st">"fire"</span>,</span>
<span>                            <span class="va">study_type</span>, <span class="va">study_suffix</span>,</span>
<span>                            <span class="va">array_identifier</span>, <span class="st">"Files/"</span>,</span>
<span>                            sep <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span><span class="co"># ftp.use.epsv ensures that the request is not rejected by the FTP web server</span></span>
<span><span class="co"># dirlistonly ensures that we only retrieve filenames from the FTP repository</span></span>
<span><span class="va">ftp_biostudies</span> <span class="op">&lt;-</span> <span class="fu">RCurl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RCurl/man/getURL.html" class="external-link">getURL</a></span><span class="op">(</span><span class="va">biostudies_ftp_url</span>,</span>
<span>                           ftp.use.epsv <span class="op">=</span> <span class="cn">FALSE</span>, dirlistonly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\r\n"</span>, simplify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="st">"[^[[:blank:]]*]"</span><span class="op">)</span> <span class="co"># remove empty chains, only composed of white spaces</span></span>
<span><span class="co"># define the connection port</span></span>
<span><span class="va">con</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/pkg/RCurl/man/getCurlHandle.html" class="external-link">getCurlHandle</a></span><span class="op">(</span>ftp.use.epsv <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># download all file listed in the FTP repository</span></span>
<span><span class="va">contents</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">filenames</span>, <span class="va">getURL</span>, curl <span class="op">=</span> <span class="va">con</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="optionnal-sra-and-ena-organisational-structure">(Optionnal) SRA and ENA organisational structure<a class="anchor" aria-label="anchor" href="#optionnal-sra-and-ena-organisational-structure"></a>
</h5>
<div class="callout-note" title="ENA hierarchical structure">
<p>The hierarchical structure in the <strong>SRA (Sequence Read
Archive)</strong> or <strong>ENA (European Nucleotide Archive)</strong>
follows the following order of granularity:</p>
<p>1.<strong>ERP (or SRP)</strong> represents a <strong>study or
project</strong>, describing the overarching goal of the sequencing
experiment and providing metadata about the research context.</p>
<ol start="2" style="list-style-type: decimal">
<li><p><strong>ERS (or SRS)</strong> represents an <strong>individual
biological sample</strong> within a study.</p></li>
<li><p><strong>ERX (or SRX)</strong> represents an
<strong>experiment</strong>, describing the methodology applied to a
sample, such as library preparation, sequencing platform and sequencing
strategy (e.g., RNA-seq, WGS).</p></li>
<li><p>The <strong>ERR (or SRR)</strong> represents a specific
<strong>run</strong>, which is the unit of raw sequencing data generated
from an experiment. Each ERR corresponds to a FASTQ file or set of FASTQ
files containing sequencing reads.</p></li>
</ol>
<p>Note that complex mappings relate these objects to each other:</p>
<ul>
<li>Multiple Experiments can be grouped together to form a Study.</li>
<li>A single Sample can be used by more than one Experiment, so there is
a many-to-many relationship between Samples and Experiments.</li>
<li>One Experiment can store multiple Runs.</li>
<li>Only the Run level has a one-to-one relationship or Paired-End,
assuming the FASTQ format was used, relationship with the original
files.</li>
</ul>
</div>
<p>More documentation on the - Main facilities delivered by the ENA web
server, including programmatic access: <a href="https://pubmed.ncbi.nlm.nih.gov/18978013/" class="external-link">Petabyte-scale
innovations at the European Nucleotide Archive</a>, from <span class="citation">Cochrane et al. (2009)</span>. - Some infographics for
detailing the interconnections between the most prominent gene
repositories are reported in Fig. <span class="citation">(<strong>fig-sra-structure?</strong>)</span>.</p>
<div id="fig-sra-structure" layout-ncol="2">
<div class="float">
<img src="img%2Fsra_databases.png" alt="Gene Expression Repositories Explained"><div class="figcaption"><a href="https://www.ccdatalab.org/blog/gene-expression-repositories-explained" class="external-link">Gene
Expression Repositories Explained</a></div>
</div>
<div class="float">
<img src="img%2Fsra_pipeline.png" alt="Pipeline to pair raw data from ENA/SRA, metadata from Biosamples and reference genoms from Ensembl and RefSeq"><div class="figcaption">Pipeline to pair raw data from ENA/SRA, metadata
from Biosamples and reference genoms from Ensembl and RefSeq</div>
</div>
<div class="float">
<img src="img%2Fsra_structure.png" alt="SRA structure"><div class="figcaption">SRA structure</div>
</div>
<div class="float">
<img src="img%2Fena_structure.jpg" alt="The relational data model for ENA reads: a study is associated with samples that are themselves related to runs. Besides, each run is associated with an experiment identifier describing which technologu has been used to perform it. Underlying this data model is an API that provides abstraction from the nature of the data file system, returning read data upon request based on read identifiers."><div class="figcaption">The relational data model for ENA reads: a study
is associated with samples that are themselves related to runs. Besides,
each run is associated with an experiment identifier describing which
technologu has been used to perform it. Underlying this data model is an
API that provides abstraction from the nature of the data file system,
returning read data upon request based on read identifiers.</div>
</div>
<p>ENA and SRA structures and layout.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 class="appendix" id="programmatic-access-to-flybase">Programmatic access to FlyBase<a class="anchor" aria-label="anchor" href="#programmatic-access-to-flybase"></a>
</h3>
<p>To retrieve datasets stored in FlyBase, such as sex, tissue, or lab
information, in an automated and programmatic manner, you can use of
these three approaches, ordered by increasing complexity but
scalability:</p>
<hr>
<div class="section level4">
<h4 id="precomputed-metadata-files">
<strong>Precomputed Metadata Files</strong>:<a class="anchor" aria-label="anchor" href="#precomputed-metadata-files"></a>
</h4>
<p>They are accessible using the <a href="https://flybase.org/downloads/bulkdata" class="external-link">FlyBase FTP</a>: - <a href="https://wiki.flybase.org/wiki/FlyBase:Downloads_Overview" class="external-link">Wiki
Documentation of the Databases structure and Overview</a></p>
<div class="callout-note" title="Only for Linux users: wget access" collapse="true">
<ol style="list-style-type: decimal">
<li>
<strong>Download the RPKM file</strong>:</li>
</ol>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">wget</span> http://ftp.flybase.org/releases/FB2024_05/precomputed_files/genes/gene_rpkm_report_fb_<span class="op">&lt;</span>version<span class="op">&gt;</span>.tsv.gz </span></code></pre></div>
<p>where <code>&lt;version&gt;</code> is a <em>placeholder</em>
referring to a FlyBase release. Alternatively, use:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">wget</span> http://ftp.flybase.org/releases/FB2024_05/precomputed_files/genes/gene_rpkm_<span class="pp">*</span>.tsv.gz </span></code></pre></div>
<p>where <code>*</code> is a regex expression standing for any symbol
(in that case, it should download two datasets).</p>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>Extract and view the file</strong>:</li>
</ol>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">gunzip</span> gene_rpkm_report_fb_2024_05.tsv.gz</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Analyze in R</strong>:</li>
</ol>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RPKM_file</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"gene_rpkm_report_fb_2024_05.tsv"</span>, </span>
<span>                             sep <span class="op">=</span> <span class="st">"\t"</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">RPKM_file</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Alternatively, we propose the following portable R script to
replicate the <code>wget</code> and <code>gunzip</code>
functionalities.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define the URL and output file paths</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"http://ftp.flybase.org/releases/FB2024_05/precomputed_files/genes/gene_rpkm_report_fb_2024_05.tsv.gz"</span></span>
<span><span class="va">output_file</span> <span class="op">&lt;-</span> <span class="st">"data/RNASeq/gene_rpkm_report_fb_2024_05.tsv.gz"</span></span>
<span><span class="va">output_unzipped</span> <span class="op">&lt;-</span> <span class="st">"data/RNASeq/gene_rpkm_report_fb_2024_05.tsv"</span></span>
<span></span>
<span><span class="co"># Step 1: Download the file</span></span>
<span><span class="va">response</span> <span class="op">&lt;-</span> <span class="fu">GET</span><span class="op">(</span><span class="va">url</span>, <span class="co"># `write_disk()` specifies path location of the download file.</span></span>
<span>                <span class="fu">write_disk</span><span class="op">(</span><span class="va">output_file</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check if download was successful</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">response</span><span class="op">$</span><span class="va">status_code</span> <span class="op">==</span> <span class="fl">200</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"File downloaded successfully to:"</span>, <span class="va">output_file</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Download failed with status code:"</span>, <span class="va">response</span><span class="op">$</span><span class="va">status_code</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Step 2: Decompress the file</span></span>
<span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu">gunzip</span><span class="op">(</span><span class="va">output_file</span>, <span class="va">output_unzipped</span>, remove <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<hr>
</div>
<div class="section level4">
<h4 id="using-the-flybase-rest-api-lightweight-approach">
<strong>Using the FlyBase REST API lightweight
approach</strong>:<a class="anchor" aria-label="anchor" href="#using-the-flybase-rest-api-lightweight-approach"></a>
</h4>
<p>The REST API allows for dynamic metadata retrieval via programmatic
queries. In particular, the <code>GET</code> request is used to retrieve
data from a server. It is however restricted by URL length. Here’s a
linux example below, using the <code>curl</code> command, and using one
parameter composed of several IDs separated by commas:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-X</span> GET <span class="st">"https://api.flybase.org/api/v1.0/hitlist/fetch/FBlc0000215,FBlc0000216,FBlc0000217"</span> <span class="at">-H</span> <span class="st">"Accept: application/json"</span></span></code></pre></div>
<p>Report to <span class="citation">(<strong>sec-modencode-metadata?</strong>)</span> for a
R high-user interface, using the complementary between the
<code>jsonlite</code> and <code>httr2</code> packages.</p>
<hr>
</div>
<div class="section level4">
<h4 id="using-the-flybase-chado-postgresql-database">
<strong>Using the FlyBase Chado PostgreSQL Database</strong><a class="anchor" aria-label="anchor" href="#using-the-flybase-chado-postgresql-database"></a>
</h4>
<p>That’s definitely the most challenging, but also the safest and most
scalable approach. In brief, you import the whole database and its
relational scheme to your local computer or server.</p>
<p>The <em>Chado database</em> is a wrapper of the
<code>PostgreSQL</code> DBMS, with custom functions to parse biological
samples.</p>
<div id="nte-chado-basic" class="callout-note" title="Use the Chado Database" collapse="true">
<ol style="list-style-type: decimal">
<li>Connect to the FlyBase public instance<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;In details, the following credentials are expected:
&lt;strong&gt;Hostname:&lt;/strong&gt;(&lt;code&gt;chado.flybase.org&lt;/code&gt;),
&lt;strong&gt;Database:&lt;/strong&gt;(&lt;code&gt;flybase&lt;/code&gt;),
&lt;strong&gt;User:&lt;/strong&gt;(&lt;code&gt;flybase&lt;/code&gt;), no password and
&lt;strong&gt;Port:&lt;/strong&gt; 5432&lt;/p&gt;"><sup>12</sup></a>.</li>
</ol>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="ex">psql</span> <span class="at">-h</span> chado.flybase.org <span class="at">-U</span> flybase flybase</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Query metadata for datasets:</li>
</ol>
<div class="sourceCode" id="cb26"><pre class="sourceCode sql"><code class="sourceCode sql"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="kw">SELECT</span> <span class="op">*</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="kw">FROM</span> dataset</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="kw">WHERE</span> dataset_id <span class="op">=</span> <span class="st">'FB2024_05'</span>;</span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Export results for further analysis:</li>
</ol>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="ex">\COPY</span> <span class="er">(</span><span class="ex">SELECT</span> <span class="pp">*</span> FROM dataset WHERE dataset_id = <span class="st">'FB2024_05'</span><span class="kw">)</span> <span class="ex">TO</span> <span class="st">'metadata.csv'</span> CSV HEADER<span class="kw">;</span></span></code></pre></div>
</div>
<div class="callout-info" title="Alternative use of the Chado Database">
<p>In <span class="citation">(<strong>nte-chado-basic?</strong>)</span>,
we described the <em>Direct Chado Query access</em>. However, this
public and read-only access, directly exploitable with an installed SQL
client, is suitable only for occasional access, and with lots of users
requesting the database, can induce a lot of latency.</p>
<p>The alternative approach consists of first <strong>downloading and
Hosting the Chado Database Locally</strong> :</p>
<ul>
<li>
<strong>Step 1:</strong> Download the PostgreSQL dump files from the
FlyBase FTP site:<br><a href="ftp://ftp.flybase.org/releases/current/psql/">FlyBase
PostgreSQL Dumps</a>.</li>
<li>
<strong>Step 2:</strong> Create a new PostgreSQL database
locally:</li>
</ul>
<div class="sourceCode" id="cb28"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="ex">createdb</span> <span class="at">-E</span> UTF-8 my_flybase</span></code></pre></div>
<ul>
<li>
<strong>Step 3:</strong> Load the downloaded data into your local
database:</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">cat</span> FB<span class="pp">*</span>.sql.gz.<span class="pp">*</span> <span class="kw">|</span> <span class="fu">gunzip</span> <span class="kw">|</span> <span class="ex">psql</span> my_flybase</span></code></pre></div>
<ul>
<li>
<strong>Step 4:</strong> Optimize the database<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;This approach is recommended for high-performance
needs.&lt;/p&gt;"><sup>13</sup></a> for use:</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="ex">vacuumdb</span> <span class="at">-f</span> <span class="at">-z</span> <span class="at">-v</span> my_flybase</span></code></pre></div>
<ul>
<li>
<strong>Chado Schema</strong>: Learn the database schema structure
to optimize queries with the <a href="https://flybase.github.io/docs/chado/functions" class="external-link">Chado
Documentation</a>.</li>
</ul>
</div>
<hr>
<table class="table">
<colgroup>
<col width="27%">
<col width="37%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th><strong>Method</strong></th>
<th><strong>Advantages</strong></th>
<th><strong>Disadvantages</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>Precomputed Metadata</strong></td>
<td>Easy to use; requires minimal setup</td>
<td>Limited to static metadata releases</td>
</tr>
<tr class="even">
<td><strong>REST API</strong></td>
<td>Dynamic queries; programmatic access</td>
<td>Requires scripting</td>
</tr>
<tr class="odd">
<td><strong>Chado PostgreSQL</strong></td>
<td>Comprehensive and direct database access</td>
<td>High setup complexity; requires SQL</td>
</tr>
</tbody>
</table>
<hr>
<ul>
<li>
<a href="https://flyatlas.gla.ac.uk/FlyAtlas2/index.html?page=help" class="external-link">Fly
Atlas 2 database</a>, then report to section
<code>Citing  FlyAtlas 2 &amp; Data Availability / Data Availability</code>.</li>
<li><a href="motif.mvls.gla.ac.uk/downloads/FlyAtlas2_2024.01.08.sql">Global
SQL databse</a></li>
<li>
<a href="https://www.ebi.ac.uk/ena/browser/view/PRJEB22205" class="external-link">Original
Fly Atlas2 datasets, including both RNASeq and Microarray</a> -&gt; to
retrieve all information, likely to use the sql databse rather than
excel, where data is aggregated/merged: <a href="https://pmc.ncbi.nlm.nih.gov/articles/instance/5753349/bin/gkx976_supp.pdf" class="external-link uri">https://pmc.ncbi.nlm.nih.gov/articles/instance/5753349/bin/gkx976_supp.pdf</a>
for an explanation of the SQL schema.</li>
<li><a href="https://github.com/YazBraimah/FlyAtlas2/blob/master/Snakefile" class="external-link">SnakeMake
GitHub repository to reproduce all the results + metadata for Fly Atlas
2</a></li>
</ul>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-cochrane2009nar" class="csl-entry">
Cochrane, Guy, Ruth Akhtar, James Bonfield, Lawrence Bower, Fehmi
Demiralp, Nadeem Faruque, Richard Gibson, et al. 2009.
<span>“Petabyte-Scale Innovations at the <span>European Nucleotide
Archive</span>.”</span> <em>Nucleic Acids Research</em> 37 (January):
D19–25. <a href="https://doi.org/10.1093/nar/gkn765" class="external-link">https://doi.org/10.1093/nar/gkn765</a>.
</div>
<div id="ref-zeeberg2004bb" class="csl-entry">
Zeeberg, Barry R., Joseph Riss, David W. Kane, Kimberly J. Bussey,
Edward Uchio, W. Marston Linehan, J. Carl Barrett, and John N.
Weinstein. 2004. <span>“Mistaken <span>Identifiers</span>:
<span>Gene</span> Name Errors Can Be Introduced Inadvertently When Using
<span>Excel</span> in Bioinformatics.”</span> <em>BMC
Bioinformatics</em> 5 (1): 80. <a href="https://doi.org/10.1186/1471-2105-5-80" class="external-link">https://doi.org/10.1186/1471-2105-5-80</a>.
</div>
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/bastienchassagnol/" class="external-link">Bastien CHASSAGNOL</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
