<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analyse single cell RNASeq experiences • DrosophiliaDeconv</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analyse single cell RNASeq experiences">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DrosophiliaDeconv</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/DrosophiliaDeconv.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/download.html">Explore and download sequencing databases: hows-to</a></li>
    <li><a class="dropdown-item" href="../articles/preprocess_bulkRNASeq.html">Analyse bulk RNASeq experiences</a></li>
    <li><a class="dropdown-item" href="../articles/preprocess_scRNASeq.html">Analyse single cell RNASeq experiences</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/bastienchassagnol/DrosophiliaDeconv/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Analyse single cell RNASeq experiences</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bastienchassagnol/DrosophiliaDeconv/blob/main/vignettes/preprocess_scRNASeq.Rmd" class="external-link"><code>vignettes/preprocess_scRNASeq.Rmd</code></a></small>
      <div class="d-none name"><code>preprocess_scRNASeq.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="relevance-proof">Relevance proof<a class="anchor" aria-label="anchor" href="#relevance-proof"></a>
</h2>
<ol style="list-style-type: decimal">
<li><a href="https://hasanyone.com/?id=623714fd" class="external-link">No deconvolutiona
glorithm used so far to infer the cellular composition from Drosophilia
samples</a></li>
<li><a href="https://typeset.io/search?q=Has%20anyone%20developped%20or%20used%20deconvolution%20algorithms%20for%20inferring%20the%20composition%20of%20bulk%20drosophilia%20samples%3F" class="external-link">With
Scispace</a></li>
</ol>
</div>
<div class="section level2">
<h2 id="recommended-scrnaseq-pipeline">Recommended scRNASeq pipeline<a class="anchor" aria-label="anchor" href="#recommended-scrnaseq-pipeline"></a>
</h2>
<ul>
<li><a href="https://github.com/omnideconv/harmony" class="external-link">Harmony and
Seurat</a></li>
<li><a href="https://github.com/omnideconv/deconvExplorer.git" class="external-link">deconvExplorer
functions to retrieve or plot marker gene sets</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="the-fca-initiative">The FCA initiative<a class="anchor" aria-label="anchor" href="#the-fca-initiative"></a>
</h2>
<ul>
<li>Original paper: <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC8944923/#SD1" class="external-link">Fly Cell
Atlas: a single-nucleus transcriptomic atlas of the adult fruit fly</a>,
from <span class="citation">Li et al. (2022)</span>.</li>
<li>
<a href="https://pmc.ncbi.nlm.nih.gov/articles/instance/8944923/bin/NIHMS1788050-supplement-FCA_supplement.pdf" class="external-link">Detailed
supplementary material</a>.</li>
</ul>
<p><img src="img%2Ffly_cell_atlas_datasets.png" alt="Types of FlyCellAtlas datasets, and how to retrieve them"> Two
web platforms have been used to automate the processing, visualisation
and downstream analyses of Drosophila samples:</p>
<ul>
<li>[SCope]((<a href="https://flycellatlas.org/scope" class="external-link uri">https://flycellatlas.org/scope</a>)</li>
<li>[ASAP]((<a href="https://flycellatlas.org/asap" class="external-link uri">https://flycellatlas.org/asap</a>)</li>
</ul>
<div class="float">
<img src="img%2Ffly_cell_atlas_preprocessing.png" alt="Key steps of the FlyCellAtlas pipeline"><div class="figcaption">Key steps of the FlyCellAtlas pipeline</div>
</div>
<ul>
<li>
<a href="https://github.com/FlyCellAtlas/data_processing" class="external-link">The
nextflow GitHub repository to process the 10X Genomic samples</a>.
However, it relies in the back-end on the <a href="https://github.com/vib-singlecell-nf/vsn-pipelines/" class="external-link">Nextflow VSP
Pipeline</a>, which is not updated anymore, preventing from simply
pre-processing data at the raw level by re-running the scripts.</li>
</ul>
</div>
<div class="section level2">
<h2 id="import-and-read-loom-and-h5ad-datasets">Import and read Loom and H5AD Datasets<a class="anchor" aria-label="anchor" href="#import-and-read-loom-and-h5ad-datasets"></a>
</h2>
<p>With <strong>Seurat</strong>, use - <code>ReadH5AD</code>: Reads
<code>.h5ad</code> files (AnnData format) or <code>ReadH5Seurat</code>:
Imports <code>.h5Seurat</code> files. - <code>ReadLoom</code>: Reads
<code>.loom</code> files for scRNA-seq data stored in the Loom
format.</p>
<p>Alternatively, you can use <strong>LoomR</strong> R package, with the
<code>read_loom_dgCMatrix</code> function (lower processing level), and
<strong>zellkonverter</strong> (for H5AD), with enhanced
interoperability between Python’s AnnData and R’s
<em>SingleCellExperiment</em> format (another lightweight alternative to
<code>Seurat</code> for simpler workflows).</p>
</div>
<div class="section level2">
<h2 id="standard-scrna-seq-pipeline-for-loomh5ad-datasets">Standard scRNA-seq Pipeline for Loom/H5AD Datasets<a class="anchor" aria-label="anchor" href="#standard-scrna-seq-pipeline-for-loomh5ad-datasets"></a>
</h2>
<div class="section level3">
<h3 id="data-import">
<strong>Data Import</strong>:<a class="anchor" aria-label="anchor" href="#data-import"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">ReadH5AD</span><span class="op">(</span><span class="st">"path/to/dataset.h5ad"</span><span class="op">)</span></span>
<span><span class="co"># or for Loom files</span></span>
<span><span class="va">loom_data</span> <span class="op">&lt;-</span> <span class="fu">ReadLoom</span><span class="op">(</span><span class="st">"path/to/dataset.loom"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quality-control-and-normalization">
<strong>Quality Control</strong> and
<strong>Normalization</strong><a class="anchor" aria-label="anchor" href="#quality-control-and-normalization"></a>
</h3>
<ul>
<li>Calculate and visualize common QC metrics (e.g., percentage of
mitochondrial genes, number of detected genes per cell).</li>
<li>Filter out poor-quality cells:</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sc_data</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span>
<span> <span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">sc_data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="feature-selection-and-scaling">
<strong>Feature Selection</strong> and scaling<a class="anchor" aria-label="anchor" href="#feature-selection-and-scaling"></a>
</h3>
<ul>
<li>Identify highly variable features (genes):</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">sc_data</span>, selection.method <span class="op">=</span> <span class="st">"vst"</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span> <span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">sc_data</span>, vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"percent.mt"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualisation-and-quality-control">visualisation and quality control<a class="anchor" aria-label="anchor" href="#visualisation-and-quality-control"></a>
</h3>
<ul>
<li>Perform PCA:</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">sc_data</span>, features <span class="op">=</span> <span class="fu">VariableFeatures</span><span class="op">(</span>object <span class="op">=</span> <span class="va">sc_data</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Perform UMAP or t-SNE for visualization:</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">RunUMAP</span><span class="op">(</span><span class="va">sc_data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Cluster the cells into distinct groups:</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">FindNeighbors</span><span class="op">(</span><span class="va">sc_data</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu">FindClusters</span><span class="op">(</span><span class="va">sc_data</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">sc_data</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="downstream-analyses">Downstream analyses<a class="anchor" aria-label="anchor" href="#downstream-analyses"></a>
</h3>
<ul>
<li><strong>Differential Expression Analysis</strong></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span><span class="va">sc_data</span>, ident.1 <span class="op">=</span> <span class="fl">1</span>, ident.2 <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<ul>
<li><strong>Integration with Other Datasets</strong></li>
</ul>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-li2022s" class="csl-entry">
Li, Hongjie, Jasper Janssens, Maxime De Waegeneer, Sai Saroja Kolluru,
Kristofer Davie, Vincent Gardeux, Wouter Saelens, et al. 2022.
<span>“Fly <span>Cell Atlas</span>: <span>A</span> Single-Nucleus
Transcriptomic Atlas of the Adult Fruit Fly.”</span> <em>Science</em>
375 (6584): eabk2432. <a href="https://doi.org/10.1126/science.abk2432" class="external-link">https://doi.org/10.1126/science.abk2432</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/bastienchassagnol/" class="external-link">Bastien CHASSAGNOL</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
